---
title: "Preliminary Analysis"
output: html_document
---

```{r setup, warning=FALSE, message=FALSE, error=FALSE}
library(tidyverse)
library(edgeR)
library(limma)
library(ggplot2)
library(dplyr)
```
***
<br><br>

## __Content__
### 1 Host expression
### 2 Symbiont expression
### 3 Next steps
### 4 Construction site (random stuff)
***
<br><br>


# __1 Host expression__
## Loading data

Read in the host count matrix and convert to log2cpm
``` {r host-data, warning=F}
host_counts <- read_tsv("data/MATRIX_Counts-FINAL-Host.txt")
host_samples <- read_tsv("data/SampleInfo_Host-STAR.txt")
annotations <- read_tsv("data/Host-Gene-Annotation.txt")

host_DGE <- DGEList(host_counts[,-1], genes = host_counts[,1]%>% bind_cols(annotations), samples = host_samples )

filt <- filterByExpr(host_DGE, design = model.matrix(~Strain, data = host_DGE$samples), min.count = 40)

host_filtered <- host_DGE[filt, , keep.lib.sizes = F]

host_filtered <- calcNormFactors(host_filtered)

host_v <- voom(host_filtered, design = model.matrix(~Strain, data = host_filtered$samples), plot = T)
```

MDS to check overall expression differences
```{r}
plotMDS(host_v, labels = host_v$targets$FileName %>% str_remove("Host\\."), col = factor( host_v$targets$FileName %>% str_replace("Host\\.(.*?)\\..*", "\\1")) %>% as.integer())
```

Fit expression model and check stats
```{r, }
host_fit <- lmFit(host_v, design = model.matrix(~ Strain, data = host_filtered$samples)) %>% 
  eBayes()

#view fit stats
topTable(host_fit)
# export toptable 
# topTable(host_fit, n=Inf) %>% write.table(, file="host_fit_table.csv", sep="\t")

#call genes as DE for different coefficients
decideTests(host_fit)

#recreate venn diagram from prior work
decideTests(host_fit)[,-1] %>% 
  vennDiagram(include=c("up","down"))
```
```{r, eval=FALSE}
#filter based on gene annotations
topTable(host_fit, number = Inf, p.value = 0.05) %>% 
  filter(EntrezGeneID_0 %>% str_detect("[H|h]eat|[G|g]lutathione|[C|c]haperone|[D|d]ismutase"))
```





Plotting 
```{r}
plot_host_gene <- function(gene) {
  expn <- host_v$E[host_v$genes$EntrezGeneID == gene, ]
  host_v$targets %>% 
    mutate(expn = expn) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    theme(legend.position = "none")
}

```
***
<br><br>

# __2 Symbiont expression__
## Loading data

Read in the symbiont count matrix and convert to log2cpm
``` {r symbiont-data, warning=F}
symbiont_counts <- read_tsv("data/MATRIX_Counts-FINAL-Symbiont.txt")
symbiont_samples <- read_tsv("data/SampleInfo_Symbiont-STAR.txt")
#annotations <- read_tsv("data/Symbiont-Gene-Annotation.txt")

symbiont_DGE <- DGEList(symbiont_counts[,-1], genes = symbiont_counts[,1], samples = symbiont_samples )

filt <- filterByExpr(symbiont_DGE, design = model.matrix(~Strain, data = symbiont_DGE$samples), min.count = 40)

symbiont_filtered <- symbiont_DGE[filt, , keep.lib.sizes = F]

symbiont_filtered <- calcNormFactors(symbiont_filtered)

symbiont_v <- voom(symbiont_filtered, design = model.matrix(~Strain, data = symbiont_filtered$samples), plot = T)
```

MDS to check overall expression differences
```{r}
plotMDS(symbiont_v, labels = symbiont_v$targets$FileName %>% str_remove("Symbiont\\."), col = factor( symbiont_v$targets$FileName %>% str_replace("Symbiont\\.(.*?)\\..*", "\\1")) %>% as.integer())
```

Fit expression model and check stats
```{r}
symbiont_fit <- lmFit(symbiont_v, design = model.matrix(~ Strain, data = symbiont_filtered$samples)) %>% 
  eBayes()

#view fit stats
topTable(symbiont_fit)
# export toptable 
# topTable(symbiont_fit, n=Inf) %>% write.table(, file="symbiont_fit_table.csv", sep="\t")

#call genes as DE for different coefficients
decideTests(symbiont_fit)

#recreate venn diagram from prior work
decideTests(symbiont_fit)[,-1] %>% 
  vennDiagram(include=c("up","down"))
```
```{r, eval=FALSE}
#filter based on gene annotations
topTable(symbiont_fit, number = Inf, p.value = 0.05) %>% 
  filter(EntrezGeneID %>% str_detect("[H|h]eat|[G|g]lutathione|[C|c]haperone|[D|d]ismutase"))
```

Plotting 
```{r}
plot_symbiont_gene <- function(gene) {
  expn <- symbiont_v$E[symbiont_v$genes$EntrezGeneID == gene, ]
  symbiont_v$targets %>% 
    mutate(expn = expn) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    theme(legend.position = "none")
}
```
***
<br><br>

# __3 Next steps__

* Summarise to gene level counts and redo DE analysis
* Investigate DEXSeq for alternate exon usage
* Annotate with GO categories for functional analysis
* Cluster expression and plot cluster patterns
* Do all the above with symbiont
* Symbiont/host interactions
* WT does produce a lot of ROS... SS dont produce ROS. What can we find in the data to support this.. looking for SOD and other ROS scavengers. May be hypothesis testing by pulling out stress related genes? What was the code again for it please?
* Looking at the heatmaps, there are signature blocks of genes that are either highly upregulated or downregul for the respective genes. What are these genes and what are their function?
* We have SS8 strain (resistant) and WT10 strain (susceptible). Minimal evolution of three years between them, yet different thermal tolerance. We should be able to pinpoint the mutations (probably SNP analysis) and modulation of pathways that contribute to the differential resistance. We really have here for the first time an apple VS apple comparison (no apple vs orange any more).
* Should we have these things as a discussion in the Git section "issues"?

***
<br><br>



# __4 Construction site__
## Heatmap of most significant genes across samples
The code below is a bit messy, sorry, I could not clean it up later. If you have a shortcut how to do the same, that would be great of course.. many thanks.
Sometimes the error comes up when running the heatmap and it fails to produce the graph (did not google it yet, will do tomorrow). But the heatmap usually works..:

Error in (function (filename = "Rplot%03d.png", width = 480, height = 480, : unable to start png() device

Oh, I think, I just realised that the heatmap is not on the logcounts, but looks like absolute values. I still need to fix that..

It may be nice to order them according to their average counts from high to low, so we would get a top block of all expressed genes for WT and a bottom block for all expressed genes for the SS strains. Will have another look into it.
***
<br><br>

Heatmap of most significant genes across samples (should be discriminant analysis, what genes distinguish WT10 from SS8?, at the moment it is across all samples).
```{r}
# Heatmap of most significant genes across samples
# Colour definitions
library(RColorBrewer)
mypalette <- brewer.pal(11, "RdYlBu")
morecols <- colorRampPalette(mypalette)
col.cell <- c("purple", "orange")[host_samples$CellType]
column_labels <- paste(host_samples$CellType, host_samples$Strain, sep = ".")
column_labels <- paste(host_samples$SampleName)

# A)
# Extract top 500 most significant genes
# Extract labels to match them later to table with logcounts
top500signf <- topTable(host_fit, n=500)
select_topsignf <- as.character(top500signf$EntrezGeneID)
head(select_topsignf)

# B) 
# Extract genes of interest that are significantly different between SS8 and WT10, sort according to logvalues (NOT WORKING)
design <- model.matrix(~Strain, data = host_filtered$samples)
colnames(design) 
cont.matrix <- makeContrasts(SS8.vs.WT10 = (Intercept) - StrainStrain8, levels=design)
cont.matrix
names(host_fit)
fit.cont <- contrasts.fit(host_fit, cont.matrix)
fit.cont <- eBayes(fit.cont)
dim(fit.cont)
summa.fit <- decideTests(fit.cont)
summary(summa.fit)
SelectedGenesSignf <- topTable(fit.cont,coef="SS8.vs.WT10", number = Inf, p.value = 0.05, sort.by="logFC") %>% 
  filter(EntrezGeneID_0 %>% str_detect("[G|g]lutathione"))
select_topsignf <- as.character(SelectedGenesSignf$EntrezGeneID)

# C)
# Extract genes of interest across all genes
SelectedGenesSignf <- topTable(host_fit, number = Inf, p.value = 0.05) %>% 
  filter(EntrezGeneID_0 %>% str_detect("[H|h]eat|[G|g]lutathione|[C|c]haperone|[D|d]ismutase"))

select_topsignf <- as.character(SelectedGenesSignf$EntrezGeneID)
head(select_topsignf)
###

## Subsetting table according to the new labels
# got some error messages that matrix is required, fixed like this
host_counts2 <- as.matrix(host_counts, mode='any', row.names=1)
# had an error that first column should have been labels, fixed like this
host_counts3 <- data.frame(host_counts2[,-1], row.names=host_counts2[,1])
# subsetting of matrix worked
highly_signf_lcpm <- host_counts3[select_topsignf, ]
# had to convert matrix into numeric matrix
highly_signf_lcpm2 <- sapply(highly_signf_lcpm, as.numeric)
dim(highly_signf_lcpm2)


library(gplots)

# make heatmap
heatmap.2(
  highly_signf_lcpm2,
  #Rowv = FALSE,
  Colv = FALSE,
  col = rev(morecols(50)),
  trace = "none",
  main = "Selected genes",
  ColSideColors = col.cell,
  scale = "row",
  labCol = column_labels,
  #fix stuff that falls off the image with the following line
  margins = c(10,5)
  # remove junk on side
  # labRow = ""
)

```


