---
title: "Preliminary Analysis"
output:
  html_document:
    df_print: paged
    toc: true
    toc_depth: 2
---

```{r setup, warning=FALSE, message=FALSE, error=FALSE}
#
# For PDF export:
# install.packages("devtools")
library(devtools)
# install_version("rmarkdown",version=1.8)
library(tidyverse)
library(edgeR)
library(limma)
library(ggplot2)
library(gplots)
library(dplyr)
library(xlsx)
library(Glimma)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
```

```{r,echo=FALSE,message=FALSE,warning=FALSE}
require(knitr)
# Set so that long lines in R will be wrapped:
opts_chunk$set(tidy.opts=list(width.cutoff=40),tidy=TRUE)
```
<br><br>



# __1 Host expression__
## Loading data

Read in the host count matrix and convert to log2cpm
``` {r host-data, message=FALSE, warning=FALSE, fig.align='center'}
host_counts <- read_tsv("data/MATRIX_Counts-FINAL-Host.txt")
host_samples <- read_tsv("data/SampleInfo_Host-STAR.txt")
host_annotations <- read_tsv("data/Aten_diamond_annot_20190117_1327.txt")
#host_exon_annnotations <- read_tsv("data/Host-Gene-Annotation.txt")
```

```{r}
# Exons
host_DGE <- DGEList(host_counts[,-1], genes = host_counts[,1], samples = host_samples)

filt <- filterByExpr(host_DGE, design = model.matrix(~Strain, data = host_DGE$samples), min.count = 40)

host_filtered1 <- host_DGE[filt, , keep.lib.sizes = F]

host_filtered2 <- calcNormFactors(host_filtered1)

host_v <- voom(host_filtered2, design = model.matrix(~Strain, data = host_filtered2$samples), plot = F)

```
<br><br>

## Collate to gene level
Might be better for some uses to look at whole gene level expression initally before drilling down to exon level data.
```{r host-gene-level-counts, fig.align='center'}
gene_host_counts <- host_counts %>% 
  separate(EntrezGeneID, c("gene", "exon"), sep="\\.exon") %>% 
  group_by(gene) %>% summarise_if(is.numeric, sum)

# check that host_annotations and genes IDs match.. needs to be all TRUE events
table((host_annotations[,1] == (gene_host_counts[,1])))

# then bind host_annotations to gene list
gene_host_DGE <- DGEList(gene_host_counts[,-1], genes = gene_host_counts[,1]%>% bind_cols(host_annotations[,-1]), samples = host_samples)

# check that annotation table and gene_host_DGE match.. needs to be all TRUE events. Two columns to match
table((gene_host_DGE$genes == (host_annotations)))

gene_host_filt <- filterByExpr(gene_host_DGE, design = model.matrix(~0 + Strain, data = gene_host_DGE$samples), min.count = 40)

gene_host_filtered1 <- gene_host_DGE[gene_host_filt, keep.lib.sizes = F]

gene_host_filtered2 <- calcNormFactors(gene_host_filtered1)


# Check filter cut off before and after
par(mfrow=c(2,2))
mean_log_cpm3 <- aveLogCPM(gene_host_DGE$counts)
filter_threshold <- 0.5
hist(mean_log_cpm3, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm3); abline(h=filter_threshold)
#
mean_log_cpm4 <- aveLogCPM(gene_host_filtered1$counts)
filter_threshold <- 0.5
hist(mean_log_cpm4, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm4); abline(h=filter_threshold)

# Check normalised and unnormalised data
par(mfrow=c(1,2))
Host_lcpm1 <- cpm(gene_host_filtered1, log=TRUE)
boxplot(Host_lcpm1, las=2, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
#
gene_host_filtered2$samples$norm.factors
Host_lcpm2 <- cpm(gene_host_filtered2, log=TRUE)
boxplot(Host_lcpm2, las=2, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")

dev.off() 


# check that data is still all matching up
table(host_samples$SampleName == colnames(gene_host_filtered2))

gene_host_v <- voom(gene_host_filtered2, design = model.matrix(~ 0 + Strain, data = gene_host_filtered2$samples), plot = T)

#
gene_host_fit_NoIntercept <- lmFit(gene_host_v, design = model.matrix(~0 +Strain, data = gene_host_filtered2$samples)) %>% eBayes()
#
plotSA(gene_host_fit_NoIntercept)

design <- model.matrix(~0 +Strain, data = gene_host_filtered2$samples)
#
colnames(design)

```
Fig. 1 Limma voom model.
<br><br>
  
## MDS to check overall expression differences
```{r, fig.align='center'}
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
plotMDS(host_v, main = "exon counts", labels = host_v$targets$FileName %>% str_remove("host\\."), col = factor(host_v$targets$Strain %>% str_replace("host\\.(.*?)\\..*", "\\1")) %>% as.integer())
#
plotMDS(gene_host_v, main = "gene counts",labels = gene_host_v$targets$FileName %>% str_remove("host\\."), col = factor(gene_host_v$targets$Strain %>% str_replace("host\\.(.*?)\\..*", "\\1")) %>% as.integer())

```
Fig. 2 MDS plot. Left side - estimate according to whole gene counts. Right side - estimate according to exon expression.
<br><br>  


## Contrasts 1
### Make contrasts to compare host SS8 to other host strains (SS3, SS5 and WT10). 
<br>
1) What separates the SS8 from all other strains? General adaptations of SS8. 
```{r Make contrasts - Host, fig.align='center'}

## CONTRASTS
## CONTRAST COMPARING THE SS8 to all other host types. 
## What are the general adaptations of SS8?
cont.matrix_host <- makeContrasts(SS8.vs.WT10 = StrainStrain8 - StrainStrain10, SS8.vs.SS3 = StrainStrain8 - StrainStrain3, SS8.vs.SS5 = StrainStrain8 - StrainStrain5, levels=design)
cont.matrix_host

## fit contrasts
fit.cont1_host <- contrasts.fit(gene_host_fit_NoIntercept, cont.matrix_host) %>% eBayes()
# fit.cont
# dim(fit.cont1host)
summary(decideTests(fit.cont1_host))

#view fit stats
topTable(fit.cont1_host)

genes_sign_SS8_Intercept_host <- decideTests(fit.cont1_host) %>% as.data.frame() %>%  bind_cols(fit.cont1_host$genes)

vennDiagram(genes_sign_SS8_Intercept[,1:3],include=c("up","down"),
            counts.col=c("brown3", "darkblue"),
            circle.col = c("darkred", "blue", "green4"))
title("Venndiagram host. SS8 strain vs all other strains [all genes]", line = 1)

```

Fig. 3 Venndiagram showing genes that are significantly different expressed between the SS8 and respective other strains. The differentially expressed genes show the general difference of SS8 to all other strains, i.e. the specific adaptations of SS8.

<br><br>

DATA EXPORT at the end of the script..

<br><br>


## Contrasts 2
<br>
What makes SS8 unique compared to all other strains? Which are the genes that are uniquely DE by SS8? 
```{r, fig.align='center'}
## CONTRAST COMPARING THE SS8 to the other host types, here WT10
cont.matrix2_host <- makeContrasts(WT10.vs.SS8 = StrainStrain10 - StrainStrain8, WT10.vs.SS3 = StrainStrain10 - StrainStrain3, WT10.vs.SS5 = StrainStrain10 - StrainStrain5, levels=design)
cont.matrix2_host

## fit contrasts
fit.cont2_host <- contrasts.fit(gene_host_fit_NoIntercept, cont.matrix2_host) %>% eBayes()

genes_sign_WT10_Intercept_host <- decideTests(fit.cont2_host) %>% as.data.frame() %>%  bind_cols(fit.cont2_host$genes)



## CONTRAST COMPARING THE SS8 to the other host types, here SS3
cont.matrix3_host <- makeContrasts(SS3.vs.SS8 = StrainStrain3 - StrainStrain8, SS3.vs.SS5 = StrainStrain3 - StrainStrain5, SS3.vs.WT10 = StrainStrain3 - StrainStrain10, levels=design)
cont.matrix3_host

## fit contrasts
fit.cont3_host <- contrasts.fit(gene_host_fit_NoIntercept, cont.matrix3_host) %>% eBayes()

genes_sign_SS3_Intercept_host <- decideTests(fit.cont3_host) %>% as.data.frame() %>%  bind_cols(fit.cont3_host$genes)



## CONTRAST COMPARING THE SS8 to the other host types, here SS5
cont.matrix4_host <- makeContrasts(SS5.vs.SS8 = StrainStrain5 - StrainStrain8, SS5.vs.SS3 = StrainStrain5 - StrainStrain3, SS5.vs.WT10 = StrainStrain5 - StrainStrain10, levels=design)
cont.matrix4_host

## fit contrasts
fit.cont4_host <- contrasts.fit(gene_host_fit_NoIntercept, cont.matrix4_host) %>% eBayes()

genes_sign_SS5_Intercept_host <- decideTests(fit.cont4_host) %>% as.data.frame() %>%  bind_cols(fit.cont4_host$genes)

```

<br><br>


## Exploratory analysis 1
### Heat map of general gene expression.
```{r DE-heatmap-plot-host, fig.align='center'}
# Filter genes called as DE to control in any strain

any_DE_genes_host <- decideTests(fit.cont1_host, lfc=1) %>% 
  as.data.frame() %>% 
  bind_cols(fit.cont2_host$genes) %>% 
  filter_if(~ is.numeric(.), any_vars(. != 0))

# Separate out ones called as DE in *all* strains as well
all_DE_genes_host <- any_DE_genes_host%>% 
  filter_if(~ is.numeric(.), all_vars(. != 0))

## Heatmap of DE genes
# Normalise expression per-gene
normalised_DE_gene_expression_host <- gene_host_v[gene_host_v$genes$gene %in% all_DE_genes_host$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(host_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup

#
# Cluster to show patterns
clustering_host <- normalised_DE_gene_expression_host %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()

#
# Plot it
expression_heatmap_host <- normalised_DE_gene_expression_host %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_host$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) + #Text
  scale_x_discrete(labels = host_samples$Strain %>% str_remove("Strain")) +
  xlab("Sample Strain")
#
expression_heatmap_host


#Can also include the actual DE calls
DE_calls_host <- any_DE_genes_host %>% 
  gather(Strain, DE_call, -gene, -Annotation) %>% 
  mutate(
    Strain = str_remove(Strain, "StrainStrain"), 
    gene = factor(gene, levels = gene %>% unique %>% .[clustering_host$order] )
    ) %>% 
  ggplot(aes(x = Strain, y = gene, fill = as.factor(DE_call))) + 
  geom_tile() + 
  theme_minimal() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(), legend.position = "none") +
  scale_x_discrete() +
  xlab("DE Calls") + 
  scale_fill_manual(values = c("blue", "white","red"))
#
cowplot::plot_grid(DE_calls_host, expression_heatmap_host + ylab(NULL), align = "v", rel_widths = c(1,5), axis = "tb")


```
<br><br>



## Exploratory analysis 2
### Hypothesis testing
<br>
Filter genes according to annotation. Which are heat stress related genes that are differentially expressed between the wild type and selected strains?
```{r, fig.align='center'}
## Filter based on gene host_annotations

# Heat stress associated genes
fit.cont1host_heat <- topTable(fit.cont1_host, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[H|h]eat|[S|s]uperoxide|[G|g]lutathione|[C|c]haperone|[P|p]eroxidase|[H|h]sp"))
fit.cont1host_heat


# Photosynthesis associated genes
fit.cont2host_apop <- topTable(fit.cont1_host, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[A|a]pop|[U|u]biquitin|[R|r]epair"))
fit.cont2host_apop


# Heat map of Top significant genes
fit.cont1host_TopSign <- topTable(fit.cont1_host, number = 100)




####
# Data feed for heat map, change as required..
heatmap.data <- fit.cont1host_heat
#

normalised_DE_gene_expression_host <- gene_host_v[gene_host_v$genes$gene %in% heatmap.data$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(host_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup
#
# Cluster to show patterns
clustering_host <- normalised_DE_gene_expression_host %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()
#
# Plot it
expression_heatmap_host <- normalised_DE_gene_expression_host %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_host$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) + # With Text
  scale_x_discrete(labels = host_samples$Strain %>% str_remove("Strain")) +
  xlab("Sample Strain")
#
expression_heatmap_host


```
<br><br>
  

## Exploratory analysis 3
### Most variable genes across samples
What are the most variable genes across samples?
```{r, message=FALSE, warning=FALSE, error=FALSE, fig.align='center'}
# Packages
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)


# Heat map of most variable genes across samples
# Take count matrix
heat.map.data <- DGEList(gene_host_filtered2$counts, samples = host_samples$Strain)
# Convert to cpm
cpm.heat <- cpm(heat.map.data, log = TRUE)
# calculate variance across gene rows
var_genes <- apply(cpm.heat, MARGIN = 1, FUN = var)
head(var_genes)
#
## Select number of genes to display and sort in a decreasing order
select_var <- names(sort(var_genes, decreasing = TRUE)) [1:500]
names(head(sort(select_var, decreasing = TRUE)))
head(select_var)
## take subsection with high logCPM
highly_variable_lcpm <- cpm.heat[select_var, ]
dim(highly_variable_lcpm)
normalised_DE_gene_expression_host <- highly_variable_lcpm

mypalette <- brewer.pal(11, "RdYlBu")
morecols <- colorRampPalette(mypalette)
col.cell <- c("purple", "orange")[host_samples$CellType]
col.Strain <- c("blue", "red", "dark green")[host_samples$Strain]
column_labels <- paste(host_samples$CellType, host_samples$Strain, sep = ".")
#column_labels <- paste(host_samples$Strain)
row_labels <- paste(gene_host_filtered2$genes)

heatmap.2(
  highly_variable_lcpm,
  #Rowv = FALSE,
  #Colv = FALSE,
  main = "Top 500 most variable genes across samples",
  ColSideColors = col.Strain,
  col = rev(morecols(50)),
  keysize = 2,
  key.par=list(mgp=c(1, 0.5, 0), mar=c(5, 2, 3, 1)),
  #lmat = lmat, 
  #lmat=rbind(c(0, 4, 2), c(0, 1, 3)), lhei=c(2.5, 10), lwid=c(1, 10, 1),
  trace = "none",
  scale = "row",
  labCol = column_labels,
  #offsetCol = 0.5,
  ## fix stuff that falls off the image with the following line
  margins = c(8,2),
  ## remove junk on side
  labRow = ""
)

```

  

  
## Plotting, plot by gene and by exon.
```{r, fig.align='center'}

## Plot by gene, according to annotation
plot_host_gene <- function(gene) {
  #expn <- gene_host_v$E[gene_host_v$genes$Annotation == gene, ]
  expn <- gene_host_v$E[gene_host_v$genes$gene == gene, ]
  gene_host_v$targets %>% 
    mutate(expn = expn) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    theme(legend.position = "none")
}

# gene of choice
par(mar=c(5,6,3,1))
layout(matrix(c(1,2), 2, 2, byrow = TRUE))

#plot_host_gene("aten_0.1.m1.446.m1")
#plot_host_gene("aten_0.1.m1.4452.m1")
plot_host_gene("aten_0.1.m1.24908.m1")
plot_host_gene("aten_0.1.m1.31125.m1")
#plot_host_gene("aten_0.1.m1.17022.m1")

dev.off() 




## Plot by exon, according to annotation
plot_host_gene_by_exon <- function(gene) {
  host_v$E %>% 
    as.data.frame() %>% 
    filter(str_detect(host_v$genes$EntrezGeneID, gene)) %>% 
    mutate(exon = host_v$genes$EntrezGeneID[str_detect(host_v$genes$EntrezGeneID, gene)]) %>% 
    gather(sample, expn, -exon) %>% 
    left_join(host_v$targets %>% as.data.frame() %>% rownames_to_column("sample")) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    facet_wrap(~factor(exon, levels = stringr::str_sort(unique(exon), numeric = T))) +
    theme(legend.position = "none")
}


# exons of gene of choice
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
#plot_host_gene_by_exon("aten_0.1.m1.446.m1")
#plot_host_gene_by_exon("aten_0.1.m1.4452.m1")
plot_host_gene_by_exon("aten_0.1.m1.24908.m1")
#plot_host_gene_by_exon("aten_0.1.m1.17022.m1")


dev.off() 

```

***
<br><br><br>



## Data export host
code silenced
```{r, eval=FALSE}

# General adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
host_SS8_intercept_VS_all <- topTable(fit.cont1_host, n=Inf) %>% write.table(, file="host_SS8_intercept_VS_all_topTable.csv", sep="\t", quote=F)

# Specific pairwise comparisons. What separates SS8 from respective strains?
library(openxlsx)
SampleGroup<-"host_SS8_intercept_VS_pairwise"
Title <- paste(SampleGroup,"DGE-Table")
FileName <- paste(Title,".xlsx")
limma.resA <- topTable(fit.cont1_host,coef="SS8.vs.WT10",sort.by="p",n="Inf")
limma.resB <- topTable(fit.cont1_host,coef="SS8.vs.SS3",sort.by="p",n="Inf")
limma.resC <- topTable(fit.cont1_host,coef="SS8.vs.SS3",sort.by="p",n="Inf")
list_of_datasets_host <- list("SS8.vs.WT10" = limma.resA, "SS8.vs.SS3" = limma.resB, "SS8.vs.SS5" = limma.resC)
write.xlsx(list_of_datasets_host, file=FileName, sep="\t", row.names=TRUE)


# General adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
host_WT10_intercept_VS_all <- topTable(fit.cont2_host, n=Inf) %>% write.table(, file="host_WT10_intercept_VS_all_topTable.csv", sep="\t", quote=F)
##
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN REVIGO
write.xlsx(genes_sign_WT10_Intercept_host, file="host_SS8_unique_DGE_expression_list_WT10.xlsx", sheetName="SS_8_vs_WT10_unique", append=TRUE, row.names=FALSE)
#
#
#
# General adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
host_SS3_intercept_VS_all <- topTable(fit.cont3_host, n=Inf) %>% write.table(, file="host_SS3_intercept_VS_all_topTable.csv", sep="\t", quote=F)
##
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN REVIGO
write.xlsx(genes_sign_SS3_Intercept_host, file="host_SS8_unique_DGE_expression_list_SS3.xlsx", sheetName="SS_8_vs_SS3_unique", append=TRUE, row.names=FALSE)
#
#
#
# General adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
host_SS5_intercept_VS_all <- topTable(fit.cont4_host, n=Inf) %>% write.table(, file="host_SS5_intercept_VS_all_topTable.csv", sep="\t", quote=F)
#
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN REVIGO
write.xlsx(genes_sign_SS5_Intercept_host, file="host_SS8_unique_DGE_expression_list_SS5.xlsx", sheetName="SS_8_vs_SS5_unique", append=T, row.names=FALSE)


```
 
<br>
<br>
<br>
  



# __2 Symbiont expression__
## Loading data

Read in the symbiont count matrix and convert to log2cpm
``` {r symbiont-data, message=FALSE, warning=FALSE, fig.align='center'}
symbiont_counts <- read_tsv("data/MATRIX_Counts-FINAL-Symbiont.txt")
symbiont_samples <- read_tsv("data/SampleInfo_Symbiont-STAR.txt")
symbiont_annotations <- read_tsv("data/Symbiont_diamond_annot_20190117_1327.txt")
#symbiont_exon_annnotations <- read_tsv("data/Symbiont-Gene-Annotation.txt")
```

```{r}
# Exons
symbiont_DGE <- DGEList(symbiont_counts[,-1], genes = symbiont_counts[,1], samples = symbiont_samples)

filt <- filterByExpr(symbiont_DGE, design = model.matrix(~Strain, data = symbiont_DGE$samples), min.count = 40)

symbiont_filtered1 <- symbiont_DGE[filt, , keep.lib.sizes = F]

symbiont_filtered2 <- calcNormFactors(symbiont_filtered1)

symbiont_v <- voom(symbiont_filtered2, design = model.matrix(~Strain, data = symbiont_filtered2$samples), plot = F)

```
<br><br>

## Collate to gene level
Might be better for some uses to look at whole gene level expression initally before drilling down to exon level data.
```{r symbiont-gene-level-counts, fig.align='center'}
gene_symbiont_counts <- symbiont_counts %>% 
  separate(EntrezGeneID, c("gene", "exon"), sep="\\.exon") %>% 
  group_by(gene) %>% summarise_if(is.numeric, sum)

# check that symbiont_annotations and genes IDs match.. needs to be all TRUE events
table((symbiont_annotations[,1] == (gene_symbiont_counts[,1])))

# then bind symbiont_annotations to gene list
gene_symbiont_DGE <- DGEList(gene_symbiont_counts[,-1], genes = gene_symbiont_counts[,1]%>% bind_cols(symbiont_annotations[,-1]), samples = symbiont_samples)

# check that annotation table and gene_symbiont_DGE match.. needs to be all TRUE events. Two columns to match
table((gene_symbiont_DGE$genes == (symbiont_annotations)))

gene_symbiont_filt <- filterByExpr(gene_symbiont_DGE, design = model.matrix(~0 + Strain, data = gene_symbiont_DGE$samples), min.count = 40)

gene_symbiont_filtered1 <- gene_symbiont_DGE[gene_symbiont_filt, keep.lib.sizes = F]

gene_symbiont_filtered2 <- calcNormFactors(gene_symbiont_filtered1)


# Check filter cut off before and after
par(mfrow=c(2,2))
mean_log_cpm3 <- aveLogCPM(gene_symbiont_DGE$counts)
filter_threshold <- 0.5
hist(mean_log_cpm3, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm3); abline(h=filter_threshold)
#
mean_log_cpm4 <- aveLogCPM(gene_symbiont_filtered1$counts)
filter_threshold <- 0.5
hist(mean_log_cpm4, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm4); abline(h=filter_threshold)

# Check normalised and unnormalised data
par(mfrow=c(1,2))
Symbiont_lcpm1 <- cpm(gene_symbiont_filtered1, log=TRUE)
boxplot(Symbiont_lcpm1, las=2, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
#
gene_symbiont_filtered2$samples$norm.factors
Symbiont_lcpm2 <- cpm(gene_symbiont_filtered2, log=TRUE)
boxplot(Symbiont_lcpm2, las=2, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")

dev.off() 


# check that data is still all matching up
table(symbiont_samples$SampleName == colnames(gene_symbiont_filtered2))

gene_symbiont_v <- voom(gene_symbiont_filtered2, design = model.matrix(~ 0 + Strain, data = gene_symbiont_filtered2$samples), plot = T)

#
gene_symbiont_fit_NoIntercept <- lmFit(gene_symbiont_v, design = model.matrix(~0 +Strain, data = gene_symbiont_filtered2$samples)) %>% eBayes()
#
plotSA(gene_symbiont_fit_NoIntercept)

design <- model.matrix(~0 +Strain, data = gene_symbiont_filtered2$samples)
#
colnames(design)

```
Fig. 1 Limma voom model.
<br><br>
  
## MDS to check overall expression differences
```{r, fig.align='center'}
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
plotMDS(symbiont_v, main = "exon counts", labels = symbiont_v$targets$FileName %>% str_remove("symbiont\\."), col = factor(symbiont_v$targets$Strain %>% str_replace("symbiont\\.(.*?)\\..*", "\\1")) %>% as.integer())
#
plotMDS(gene_symbiont_v, main = "gene counts",labels = gene_symbiont_v$targets$FileName %>% str_remove("symbiont\\."), col = factor(gene_symbiont_v$targets$Strain %>% str_replace("symbiont\\.(.*?)\\..*", "\\1")) %>% as.integer())

```
Fig. 2 MDS plot. Left side - estimate according to whole gene counts. Right side - estimate according to exon expression.
<br><br>  


## Contrasts 1
### Make contrasts to compare symbiont SS8 to other symbiont strains (SS3, SS5 and WT10). 
<br>
1) What separates the SS8 from all other strains? General adaptations of SS8. 
```{r Make contrasts - Symbiont, fig.align='center'}

## CONTRASTS
## CONTRAST COMPARING THE SS8 to all other symbiont types. 
## What are the general adaptations of SS8?
cont.matrix_sym <- makeContrasts(SS8.vs.WT10 = StrainStrain8 - StrainStrain10, SS8.vs.SS3 = StrainStrain8 - StrainStrain3, SS8.vs.SS5 = StrainStrain8 - StrainStrain5, levels=design)
cont.matrix_sym

## fit contrasts
fit.cont1_symbiont <- contrasts.fit(gene_symbiont_fit_NoIntercept, cont.matrix_sym) %>% eBayes()
# fit.cont
# dim(fit.cont1symbiont)
summary(decideTests(fit.cont1_symbiont))

#view fit stats
topTable(fit.cont1_symbiont)

genes_sign_SS8_Intercept_symbiont <- decideTests(fit.cont1_symbiont) %>% as.data.frame() %>%  bind_cols(fit.cont1_symbiont$genes)

vennDiagram(genes_sign_SS8_Intercept[,1:3],include=c("up","down"),
            counts.col=c("brown3", "darkblue"),
            circle.col = c("darkred", "blue", "green4"))
title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)

```

Fig. 3 Venndiagram showing genes that are significantly different expressed between the SS8 and respective other strains. The differentially expressed genes show the general difference of SS8 to all other strains, i.e. the specific adaptations of SS8.

<br><br>

DATA EXPORT at the end of the script..

<br><br>


## Contrasts 2
<br>
What makes SS8 unique compared to all other strains? Which are the genes that are uniquely DE by SS8? 
```{r, fig.align='center'}
## CONTRAST COMPARING THE SS8 to the other symbiont types, here WT10
cont.matrix2_symbiont <- makeContrasts(WT10.vs.SS8 = StrainStrain10 - StrainStrain8, WT10.vs.SS3 = StrainStrain10 - StrainStrain3, WT10.vs.SS5 = StrainStrain10 - StrainStrain5, levels=design)
cont.matrix2_symbiont

## fit contrasts
fit.cont2_symbiont <- contrasts.fit(gene_symbiont_fit_NoIntercept, cont.matrix2_symbiont) %>% eBayes()

genes_sign_WT10_Intercept_symbiont <- decideTests(fit.cont2_symbiont) %>% as.data.frame() %>%  bind_cols(fit.cont2_symbiont$genes)



## CONTRAST COMPARING THE SS8 to the other symbiont types, here SS3
cont.matrix3_symbiont <- makeContrasts(SS3.vs.SS8 = StrainStrain3 - StrainStrain8, SS3.vs.SS5 = StrainStrain3 - StrainStrain5, SS3.vs.WT10 = StrainStrain3 - StrainStrain10, levels=design)
cont.matrix3_symbiont

## fit contrasts
fit.cont3_symbiont <- contrasts.fit(gene_symbiont_fit_NoIntercept, cont.matrix3_symbiont) %>% eBayes()

genes_sign_SS3_Intercept_symbiont <- decideTests(fit.cont3_symbiont) %>% as.data.frame() %>%  bind_cols(fit.cont3_symbiont$genes)



## CONTRAST COMPARING THE SS8 to the other symbiont types, here SS5
cont.matrix4_symbiont <- makeContrasts(SS5.vs.SS8 = StrainStrain5 - StrainStrain8, SS5.vs.SS3 = StrainStrain5 - StrainStrain3, SS5.vs.WT10 = StrainStrain5 - StrainStrain10, levels=design)
cont.matrix4_symbiont

## fit contrasts
fit.cont4_symbiont <- contrasts.fit(gene_symbiont_fit_NoIntercept, cont.matrix4_symbiont) %>% eBayes()

genes_sign_SS5_Intercept_symbiont <- decideTests(fit.cont4_symbiont) %>% as.data.frame() %>%  bind_cols(fit.cont4_symbiont$genes)

```

<br><br>


## Exploratory analysis 1
### Heat map of general gene expression.
```{r DE-heatmap-plot-symbiont, fig.align='center'}
# Filter genes called as DE to control in any strain

any_DE_genes_symbiont <- decideTests(fit.cont1_symbiont, lfc=0.5) %>% 
  as.data.frame() %>% 
  bind_cols(fit.cont2_symbiont$genes) %>% 
  filter_if(~ is.numeric(.), any_vars(. != 0))

# Separate out ones called as DE in *all* strains as well
all_DE_genes_symbiont <- any_DE_genes_symbiont%>% 
  filter_if(~ is.numeric(.), all_vars(. != 0))

## Heatmap of DE genes
# Normalise expression per-gene
normalised_DE_gene_expression_symbiont <- gene_symbiont_v[gene_symbiont_v$genes$gene %in% all_DE_genes_symbiont$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(symbiont_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup

#
# Cluster to show patterns
clustering_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()

#
# Plot it
expression_heatmap_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_symbiont$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_text(size=8)) + #Text
  scale_x_discrete(labels = symbiont_samples$Strain %>% str_remove("Strain")) +
  xlab("Sample Strain")
#
expression_heatmap_symbiont


#Can also include the actual DE calls
DE_calls_symbiont <- any_DE_genes_symbiont %>% 
  gather(Strain, DE_call, -gene, -Annotation) %>% 
  mutate(
    Strain = str_remove(Strain, "StrainStrain"), 
    gene = factor(gene, levels = gene %>% unique %>% .[clustering_symbiont$order] )
    ) %>% 
  ggplot(aes(x = Strain, y = gene, fill = as.factor(DE_call))) + 
  geom_tile() + 
  theme_minimal() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(), legend.position = "none") +
  scale_x_discrete() +
  xlab("DE Calls") + 
  scale_fill_manual(values = c("blue", "white","red"))
#
cowplot::plot_grid(DE_calls_symbiont, expression_heatmap_symbiont + ylab(NULL), align = "v", rel_widths = c(1,5), axis = "tb")


```
<br><br>



## Exploratory analysis 2
### Hypothesis testing
<br>
Filter genes according to annotation. Which are heat stress related genes that are differentially expressed between the wild type and selected strains?
```{r, fig.align='center'}
## Filter based on gene symbiont_annotations

# Heat stress associated genes
fit.cont1symbiont_heat <- topTable(fit.cont1_symbiont, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[H|h]eat|[S|s]uperoxide|[G|g]lutathione|[C|c]haperone|[P|p]eroxidase|[H|h]sp"))
fit.cont1symbiont_heat


# Photosynthesis associated genes
fit.cont2symbiont_photo <- topTable(fit.cont1_symbiont, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[R|r]ibulose|[P|p]hoto|[F|f]ucox|[C|c]aroteno"))
fit.cont2symbiont_photo

# Heat map of Top significant genes
fit.cont1symbiont_TopSign <- topTable(fit.cont1_symbiont, number = 100)




####
# Data feed for heat map, change as required..
heatmap.data <- fit.cont2symbiont_photo
#

normalised_DE_gene_expression_symbiont <- gene_symbiont_v[gene_symbiont_v$genes$gene %in% heatmap.data$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(symbiont_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup
#
# Cluster to show patterns
clustering_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()
#
# Plot it
expression_heatmap_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_symbiont$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_text(size=8)) + # With Text
  scale_x_discrete(labels = symbiont_samples$Strain %>% str_remove("Strain")) +
  xlab("Sample Strain")
#
expression_heatmap_symbiont


```
<br><br>
  

## Exploratory analysis 3
### Most variable genes across samples
What are the most variable genes across samples?
```{r, message=FALSE, warning=FALSE, error=FALSE, fig.align='center'}
# Packages
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)


# Heat map of most variable genes across samples
# Take count matrix
heat.map.data <- DGEList(gene_symbiont_filtered2$counts, samples = symbiont_samples$Strain)
# Convert to cpm
cpm.heat <- cpm(heat.map.data, log = TRUE)
# calculate variance across gene rows
var_genes <- apply(cpm.heat, MARGIN = 1, FUN = var)
head(var_genes)
#
## Select number of genes to display and sort in a decreasing order
select_var <- names(sort(var_genes, decreasing = TRUE)) [1:50]
names(head(sort(select_var, decreasing = TRUE)))
head(select_var)
## take subsection with high logCPM
highly_variable_lcpm <- cpm.heat[select_var, ]
dim(highly_variable_lcpm)
normalised_DE_gene_expression_symbiont <- highly_variable_lcpm

mypalette <- brewer.pal(11, "RdYlBu")
morecols <- colorRampPalette(mypalette)
col.cell <- c("purple", "orange")[symbiont_samples$CellType]
col.Strain <- c("blue", "red", "dark green")[symbiont_samples$Strain]
column_labels <- paste(symbiont_samples$CellType, symbiont_samples$Strain, sep = ".")
#column_labels <- paste(symbiont_samples$Strain)
row_labels <- paste(gene_symbiont_filtered2$genes)

heatmap.2(
  highly_variable_lcpm,
  #Rowv = FALSE,
  #Colv = FALSE,
  main = "Top 500 most variable genes across samples",
  ColSideColors = col.Strain,
  col = rev(morecols(50)),
  keysize = 2,
  key.par=list(mgp=c(1, 0.5, 0), mar=c(5, 2, 3, 1)),
  #lmat = lmat, 
  #lmat=rbind(c(0, 4, 2), c(0, 1, 3)), lhei=c(2.5, 10), lwid=c(1, 10, 1),
  trace = "none",
  scale = "row",
  labCol = column_labels,
  #offsetCol = 0.5,
  ## fix stuff that falls off the image with the following line
  margins = c(8,2),
  ## remove junk on side
  labRow = ""
)

```

  

  
## Plotting, plot by gene and by exon.
```{r, fig.align='center'}

## Plot by gene, according to annotation
plot_symbiont_gene <- function(gene) {
  #expn <- gene_symbiont_v$E[gene_symbiont_v$genes$Annotation == gene, ]
  expn <- gene_symbiont_v$E[gene_symbiont_v$genes$gene == gene, ]
  gene_symbiont_v$targets %>% 
    mutate(expn = expn) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    theme(legend.position = "none")
}

# gene of choice
par(mar=c(5,6,3,1))
layout(matrix(c(1,2), 2, 2, byrow = TRUE))

plot_symbiont_gene("SymbC1.scaffold7634.1.m1")  # Superoxide dismutase, SS8 UNIQUE
#plot_symbiont_gene("SymbC1.scaffold10965.1.m1") # family chaperone, SS8 UNIQUE
#plot_symbiont_gene("SymbC1.scaffold213.10.m1")  # oxidoreductase activity, SS8 UNIQUE
#plot_symbiont_gene("SymbC1.scaffold2883.4.m1")  # Lactoylglutathione lyase
#plot_symbiont_gene("SymbC1.scaffold714.1.m1")   # Chaperone chloroplastic
#plot_symbiont_gene("SymbC1.scaffold1286.6.m1")  #	Glutathione synthetase
#plot_symbiont_gene("SymbC1.scaffold2650.2.m1")  #	E3 ubiquitin- ligase HERC1
#plot_symbiont_gene("SymbC1.scaffold12561.1.m1") #	Cytochrome mitochondrial
plot_symbiont_gene("SymbC1.scaffold9320.1.m1")  # Heat shock 70 kDa 


plot_symbiont_gene("SymbC1.scaffold8083.1.m1")  

dev.off() 




## Plot by exon, according to annotation
plot_symbiont_gene_by_exon <- function(gene) {
  symbiont_v$E %>% 
    as.data.frame() %>% 
    filter(str_detect(symbiont_v$genes$EntrezGeneID, gene)) %>% 
    mutate(exon = symbiont_v$genes$EntrezGeneID[str_detect(symbiont_v$genes$EntrezGeneID, gene)]) %>% 
    gather(sample, expn, -exon) %>% 
    left_join(symbiont_v$targets %>% as.data.frame() %>% rownames_to_column("sample")) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    facet_wrap(~factor(exon, levels = stringr::str_sort(unique(exon), numeric = T))) +
    theme(legend.position = "none")
}


# exons of gene of choice
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
plot_symbiont_gene_by_exon("SymbC1.scaffold7634.1.m1")    # Superoxide dismutase, SS8 UNIQUE
#plot_symbiont_gene_by_exon("SymbC1.scaffold10965.1.m1")   # family chaperone, SS8 UNIQUE UNIQUE
#plot_symbiont_gene_by_exon("SymbC1.scaffold213.10.m1")    # oxidoreductase activity, SS8 UNIQUE
#plot_symbiont_gene_by_exon("SymbC1.scaffold2883.4.m1")    # Lactoylglutathione lyase
#plot_symbiont_gene_by_exon("SymbC1.scaffold714.1.m1")     # Chaperone chloroplastic
plot_symbiont_gene_by_exon("SymbC1.scaffold1787.3.m1")    # hsp organising protein
plot_symbiont_gene_by_exon("SymbC1.scaffold2730.1.m1")  


dev.off() 

```

***
<br><br><br>



## Data export symbiont
code silenced
```{r, eval=FALSE}

# Gerneral adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
symbiont_SS8_intercept_VS_all <- topTable(fit.cont1_symbiont, n=Inf) %>% write.table(, file="symbiont_SS8_intercept_VS_all_topTable.csv", sep="\t", quote=F)
#
# Specific pairwise comparisons. What separates SS8 from respective strains?
library(openxlsx)
SampleGroup<-"symbiont_SS8_intercept_VS_pairwise"
Title <- paste(SampleGroup,"DGE-Table")
FileName <- paste(Title,".xlsx")
limma.resA <- topTable(fit.cont1_symbiont,coef="SS8.vs.WT10",sort.by="p",n="Inf")
limma.resB <- topTable(fit.cont1_symbiont,coef="SS8.vs.SS3",sort.by="p",n="Inf")
limma.resC <- topTable(fit.cont1_symbiont,coef="SS8.vs.SS5",sort.by="p",n="Inf")
list_of_datasets_symbiont <- list("SS8.vs.WT10" = limma.resA, "SS8.vs.SS3" = limma.resB, "SS8.vs.SS5" = limma.resC)
write.xlsx(list_of_datasets_symbiont, file=FileName, sep="\t", row.names=TRUE)
#

#
# Gerneral adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
symbiont_WT10_intercept_VS_all <- topTable(fit.cont2_symbiont, n=Inf) %>% write.table(, file="symbiont_WT10_intercept_VS_all_topTable.csv", sep="\t", quote=F)
##
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN REVIGO
write.xlsx(genes_sign_WT10_Intercept_symbiont, file="symbiont_SS8_unique_DGE_expression_list_WT10.xlsx", sheetName="SS_8_vs_WT10_unique", append=TRUE, row.names=FALSE)
#
#
#
# Gerneral adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
symbiont_SS3_intercept_VS_all <- topTable(fit.cont3_symbiont, n=Inf) %>% write.table(, file="symbiont_SS3_intercept_VS_all_topTable.csv", sep="\t", quote=F)
##
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN REVIGO
write.xlsx(genes_sign_SS3_Intercept_symbiont, file="symbiont_SS8_unique_DGE_expression_list_SS3.xlsx", sheetName="SS_8_vs_SS3_unique", append=TRUE, row.names=FALSE)
#
#
#
# Gerneral adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
symbiont_SS5_intercept_VS_all <- topTable(fit.cont4_symbiont, n=Inf) %>% write.table(, file="symbiont_SS5_intercept_VS_all_topTable.csv", sep="\t", quote=F)
#
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN REVIGO
write.xlsx(genes_sign_SS5_Intercept_symbiont, file="symbiont_SS8_unique_DGE_expression_list_SS5.xlsx", sheetName="SS_8_vs_SS5_unique", append=TRUE, row.names=FALSE)



```
 
<br>
<br>
<br>
  





# __3 Next steps__

* Investigate DEXSeq for alternate exon usage
    - needs to be done

* Cluster expression / WGCNA and plot cluster patterns
    - needs to be done, Stephen?
  
* Symbiont/host interactions
    - MixOmics package, needs to be done
    
* SNP analysis

* Different copies of genes used, alignment and cmparison


<br>

***
<br><br>



# __4 Extra Code__
## Heat map from external matrix

Extra chunk of code to produce heat map from external matrix (silenced).
```{r, eval=FALSE}

# HEAT MAP FROM IMPORTED MATRIX
#SampleGroup<-"Host-STAR"
time<-format(Sys.time(),"%d-%m-%Y")
#
# LOAD MATRIX
# Only heat stress realted genes
highly_variable_lcpm_NEW <- read.delim("data/Heatmaps/Host-NewFilter-Apoptosis_SS3_2.txt", row.names=1)
highly_variable_lcpm_NEWNEW <- as.matrix(highly_variable_lcpm_NEW, mode='any')
#
#
# SELECTION IN R DOES NOT WORK - see var_genes file is a value with attributes, not sure how to convert TEST2NEW to it in order the subsetting would work..
#library(dplyr)
#TEST <- topTable(fit.cont, n=Inf, coef="SS8.vs.WT10")
#TEST2 <- subset(TEST, select = c(5))
#head(TEST2)
#TEST2NEW <- as.matrix(TEST2, mode='numeric')
#select_var_NEW <- names(sort(TEST2NEW, decreasing = TRUE)) [1:500]
#head(select_var_NEW)
#
# Make new heat map
TitleFile <- paste(SampleGroup,"Apop")
Title <- paste(SampleGroup,"- stress realted genes - sign <0.0001 genes across all samples")
FileName <- paste(TitleFile,time,".jpg")
jpeg(file=FileName, width = 4000, height = 3280, res = 450)
par(mfrow = c(1,1))
#
lmat = rbind(0:3,2:1,4:4)
#lmat = rbind(c(0,3),c(2,1),c(0,4))
lwid = c(1.5,4)
lhei = c(1.5,4,1)
# lmat
#

mypalette <- brewer.pal(11, "RdYlBu")
morecols <- colorRampPalette(mypalette)
col.cell <- c("purple", "orange")[host_samples$CellType]
column_labels <- paste(host_samples$CellType, host_samples$Strain, sep = ".")
column_labels <- paste(host_samples$SampleName)

#
heatmap.2(
  highly_variable_lcpm_NEWNEW,
  Rowv = FALSE,
  Colv = FALSE,
  col = rev(morecols(50)),
  trace = "none",
  key.title = NA,
  keysize = 1,
  key.par=list(mgp=c(1, 0.5, 0), mar=c(5, 2, 2, 1)),
  #lmat = lmat, 
  #lmat=rbind(c(0, 4, 2), c(0, 1, 3)), lhei=c(2.5, 10), lwid=c(1, 10, 1),
  main = " ",
  ColSideColors = col.cell,
  scale = "row",
  labCol = column_labels,
  ## fix stuff that falls off the image with the following line
  margins = c(8,32)
  ## remove junk on side
  #labRow = ""
)
title(Title, line = +3.0, adj = +0.4)
dev.off() 

```
<br><br>







## MDS Statistics with Adonis - Host
Test Code - MDS stats Adonis
```{r, fig.align='center'}
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)

#
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
#
## MDS PLOT FROM ANY MATRIX
## MDS PLOT FROM ANY MATRIX
## MDS PLOT FROM ANY MATRIX


# DIFFERENT INPUT - USE logCPM on counts before voom
#head(gene_host_filtered1$counts)
#head(logcpmtmm)
#logcounts <- gene_host_filtered1$counts
  
#head(gene_host_filtered2$counts)
#logcounts <- gene_host_filtered2$counts

myCPM_host <- cpm(gene_host_filtered2$counts, log = TRUE)
#head(myCPM_host)
logcounts_host <- myCPM_host

#head(gene_host_v$E)
#logcounts <- gene_host_v$E
#head(logcounts)


#
# Make Factors
CellType=as.character(host_samples$CellType)
Strain=as.character(host_samples$Strain)
#CellType
#Strain
conditions=data.frame(cbind(CellType,Strain))
#conditions
#
#
# Calculate distances etc..
dds.pcoa3=pcoa(vegdist(t(logcounts_host),method="euclidean")/1000)
# export values
scores3=dds.pcoa3$vectors
#
#head(scores3)
#
#
SampleGroup<-"Host-STAR"
#SampleGroup<-"Symb-STAR"
time<-format(Sys.time(),"%d-%m-%Y")
#
Title <- paste(SampleGroup,"MDS-logcounts-euclidean")
FileName <- paste(Title,time,".jpg")
jpeg(file=FileName, width = 2020, height = 1080, res = 200)

par(mfrow=c(1,2))
#
plot(scores3[,1], scores3[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","limegreen","limegreen","limegreen","limegreen","limegreen","Red","Red","Red","Red","Red","Red"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3,CellType,label=T)
ordiellipse(scores3,CellType)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)
#
plot(scores3[,1], scores3[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","limegreen","limegreen","limegreen","limegreen","limegreen","Red","Red","Red","Red","Red","Red"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3,Strain,label=T)
ordiellipse(scores3,Strain)
#
Title <- paste(SampleGroup,"MDS-Plot")
#
title(Title)
dev.off() 

## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(CellType)),type="s",radius=0.005)

##
## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(Strain)),type="s",radius=0.005)

##
##
##

## STATISTIC TESTS - Adonis
## STATISTIC TESTS - Adonis
## STATISTIC TESTS - Adonis
##
Title <- paste(SampleGroup,"Adonis")
FileName <- paste("5-",Title,time,".txt")
ad_host=adonis(t(logcounts_host)~CellType+Strain, permutations = 999, data=conditions,method="euclidean")
ad_host 
#
#capture.output(ad_host,file=FileName)
#
Title <- paste(SampleGroup,"TukeyHSD")
FileName <- paste("5-",Title,time,".txt")
ad_host=adonis(t(logcounts_host)~Strain,data=conditions,method="euclidean")
ad_host
pco1=scores3[,1]
TukeyHSD(aov(pco1~Strain))
#
#capture.output(TukeyHSD(aov(pco1~Strain)),file=FileName)

```


<br><br>





## MDS Statistics with Adonis - Symbiont
Test Code - MDS stats Adonis
```{r, fig.align='center'}
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)

#
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
######## MDS-WITH ADONIS
#
## MDS PLOT FROM ANY MATRIX
## MDS PLOT FROM ANY MATRIX
## MDS PLOT FROM ANY MATRIX

# DIFFERENT INPUT - USE logCPM on counts before voom
#head(gene_symbiont_filtered1$counts)
#head(logcpmtmm)
#logcounts <- gene_symbiont_filtered1$counts
  
#head(gene_symbiont_filtered2$counts)
#logcounts <- gene_symbiont_filtered2$counts

myCPM_symbiont <- cpm(gene_symbiont_filtered2$counts, log = TRUE)
#head(myCPM_symbiont)
logcounts_symbiont <- myCPM_symbiont

#head(gene_symbiont_v$E)
#logcounts <- gene_symbiont_v$E
#head(logcounts)


#
# Make Factors
CellType=as.character(symbiont_samples$CellType)
Strain=as.character(symbiont_samples$Strain)
#CellType
#Strain
conditions=data.frame(cbind(CellType,Strain))
#conditions
#
# Calculate distances etc..
dds.pcoa3=pcoa(vegdist(t(logcounts_symbiont),method="euclidean")/1000)
# export values
scores3=dds.pcoa3$vectors
#
# head(scores3)
#
#
#SampleGroup<-"Host-STAR"
SampleGroup<-"Symb-STAR"
time<-format(Sys.time(),"%d-%m-%Y")
#
#Title <- paste(SampleGroup,"MDS-logcounts-euclidean-logCPM")
#FileName <- paste(Title,time,".jpg")
#jpeg(file=FileName, width = 2020, height = 1080, res = 200)
#par(mfrow=c(1,2))
#
plot(scores3[,1], scores3[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","limegreen","limegreen","limegreen","limegreen","limegreen","Red","Red","Red","Red","Red","Red"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3,CellType,label=T)
ordiellipse(scores3,CellType)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)
#
plot(scores3[,1], scores3[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","limegreen","limegreen","limegreen","limegreen","limegreen","Red","Red","Red","Red","Red","Red"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3,Strain,label=T)
ordiellipse(scores3,Strain)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)
#
# dev.off() 

## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(CellType)),type="s",radius=0.005)

##
## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(Strain)),type="s",radius=0.005)

##
##
##

## STATISTIC TESTS - Adonis
## STATISTIC TESTS - Adonis
## STATISTIC TESTS - Adonis
##
Title <- paste(SampleGroup,"Adonis")
FileName <- paste("5-",Title,time,".txt")
#
ad_symbiont=adonis(t(logcounts_symbiont)~CellType+Strain, permutations = 999, data=conditions,method="euclidean")
ad_symbiont 
#
#capture.output(ad_symbiont,file=FileName)
#
Title <- paste(SampleGroup,"TukeyHSD")
FileName <- paste("5-",Title,time,".txt")
ad_symbiont=adonis(t(logcounts_symbiont)~Strain,data=conditions,method="euclidean")
ad_symbiont
pco1=scores3[,1]
TukeyHSD(aov(pco1~Strain))
#
#capture.output(TukeyHSD(aov(pco1~Strain)),file=FileName)

```




