---
title: "R Notebook - Directed evolution of algal endosymbionts can enhance coral thermal
  tolerance"
output:
  html_document:
    df_print: paged
    toc: true
    toc_depth: 2
---

<br>

***
### Code for gene expression analysis
***

<br>


```{r setup, warning=FALSE, message=FALSE, error=FALSE}
#
# For PDF export:
# install.packages("devtools")
library(devtools)
# install_version("rmarkdown",version=1.8)
library(tidyverse)
library(edgeR)
library(limma)
library(ggplot2)
library(gplots)
library(dplyr)
library(xlsx)
library(Glimma)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
```

```{r,echo=FALSE,message=FALSE,warning=FALSE}
require(knitr)
# Set so that long lines in R will be wrapped:
opts_chunk$set(tidy.opts=list(width.cutoff=40),tidy=TRUE)
```
<br><br>



# __1 Host expression__
## Loading data

Read in the host count matrix and convert to log2cpm
``` {r host-data, message=FALSE, warning=FALSE, fig.align='center'}
host_counts <- read_tsv("data/MATRIX_Counts-FINAL-Host.txt")
host_samples <- read_tsv("data/SampleInfo_Host-STAR.txt")
host_annotations <- read_tsv("data/Aten_diamond_annot_20190117_1327.txt")
#host_exon_annnotations <- read_tsv("data/Host-Gene-Annotation.txt")
```

```{r}
# Exons
host_DGE <- DGEList(host_counts[,-1], genes = host_counts[,1], samples = host_samples)

filt <- filterByExpr(host_DGE, design = model.matrix(~Strain, data = host_DGE$samples), min.count = 40)

host_filtered1 <- host_DGE[filt, , keep.lib.sizes = F]

host_filtered2 <- calcNormFactors(host_filtered1)

host_v <- voom(host_filtered2, design = model.matrix(~Strain, data = host_filtered2$samples), plot = F)

```
<br><br>

## Collate to gene level
Might be better for some uses to look at whole gene level expression initally before drilling down to exon level data.
```{r host-gene-level-counts, fig.align='center'}
gene_host_counts <- host_counts %>% 
  separate(EntrezGeneID, c("gene", "exon"), sep="\\.exon") %>% 
  group_by(gene) %>% summarise_if(is.numeric, sum)

# check that host_annotations and genes IDs match.. needs to be all TRUE events
table((host_annotations[,1] == (gene_host_counts[,1])))

# then bind host_annotations to gene list
gene_host_DGE <- DGEList(gene_host_counts[,-1], genes = gene_host_counts[,1]%>% bind_cols(host_annotations[,-1]), samples = host_samples)

# check that annotation table and gene_host_DGE match.. needs to be all TRUE events. Two columns to match
table((gene_host_DGE$genes == (host_annotations)))

gene_host_filt <- filterByExpr(gene_host_DGE, design = model.matrix(~0 + Strain, data = gene_host_DGE$samples), min.count = 40)

gene_host_filtered1 <- gene_host_DGE[gene_host_filt, keep.lib.sizes = F]

gene_host_filtered2 <- calcNormFactors(gene_host_filtered1)


# Check filter cut off before and after
par(mfrow=c(2,2))
mean_log_cpm3 <- aveLogCPM(gene_host_DGE$counts)
filter_threshold <- 0.5
hist(mean_log_cpm3, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm3); abline(h=filter_threshold)
#
mean_log_cpm4 <- aveLogCPM(gene_host_filtered1$counts)
filter_threshold <- 0.5
hist(mean_log_cpm4, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm4); abline(h=filter_threshold)

# Check normalised and unnormalised data
par(mfrow=c(1,2))
Host_lcpm1 <- cpm(gene_host_filtered1, log=TRUE)
boxplot(Host_lcpm1, las=2, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
#
gene_host_filtered2$samples$norm.factors
Host_lcpm2 <- cpm(gene_host_filtered2, log=TRUE)
boxplot(Host_lcpm2, las=2, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")


# check that data is still all matching up
table(host_samples$SampleName == colnames(gene_host_filtered2))

gene_host_v <- voom(gene_host_filtered2, design = model.matrix(~ 0 + Strain, data = gene_host_filtered2$samples), plot = T)

#
gene_host_fit_NoIntercept <- lmFit(gene_host_v, design = model.matrix(~0 +Strain, data = gene_host_filtered2$samples)) %>% eBayes()
#
plotSA(gene_host_fit_NoIntercept)

design <- model.matrix(~0 +Strain, data = gene_host_filtered2$samples)
#
colnames(design)

```
Fig. 1 Limma voom model.
<br><br>
  
## MDS to check overall expression differences
```{r, fig.align='center'}
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
plotMDS(host_v, main = "exon counts", labels = host_v$targets$FileName %>% str_remove("host\\."), col = factor(host_v$targets$Strain %>% str_replace("host\\.(.*?)\\..*", "\\1")) %>% as.integer())
#
plotMDS(gene_host_v, main = "gene counts",labels = gene_host_v$targets$FileName %>% str_remove("host\\."), col = factor(gene_host_v$targets$Strain %>% str_replace("host\\.(.*?)\\..*", "\\1")) %>% as.integer())

```
Fig. 2 MDS plot.

<br><br>  


## Contrasts 1
<br>
```{r Make contrasts - Host, fig.align='center'}

## CONTRASTS
## CONTRAST COMPARING THE SS8 to all other host types. 
## What are the general adaptations of SS8?
cont.matrix_host <- makeContrasts(SS8.vs.WT10 = StrainStrain8 - StrainStrain10, SS8.vs.SS3 = StrainStrain8 - StrainStrain3, SS8.vs.SS5 = StrainStrain8 - StrainStrain5, levels=design)
cont.matrix_host

## fit contrasts
fit.cont1_host <- contrasts.fit(gene_host_fit_NoIntercept, cont.matrix_host) %>% eBayes()
# fit.cont
# dim(fit.cont1host)
summary(decideTests(fit.cont1_host))

#view fit stats
topTable(fit.cont1_host)

genes_sign_SS8_Intercept_host <- decideTests(fit.cont1_host) %>% as.data.frame() %>%  bind_cols(fit.cont1_host$genes)

vennDiagram(genes_sign_SS8_Intercept_host[,1:3],include=c("up","down"),
            counts.col=c("brown3", "darkblue"),
            circle.col = c("darkred", "blue", "green4"))
title("Venndiagram host. SS8 strain vs all other strains [all genes]", line = 1)


#genes_sign_SS8_Intercept_host2 <- decideTests(fit.cont1_host, lfc=1) %>% as.data.frame() %>%  bind_cols(fit.cont1_host$genes)
#vennDiagram(genes_sign_SS8_Intercept_host2[,1:3],include=c("up","down"),
#            counts.col=c("brown3", "darkblue"),
#            circle.col = c("darkred", "blue", "green4"))
#title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)

```

Fig. 3 Venndiagram showing genes that are significantly different expressed between the SS8 and respective other strains. The differentially expressed genes show the general difference of SS8 to all other strains, i.e. the specific adaptations of SS8.

<br><br>


## Contrasts 2
<br>
What makes SS8 unique compared to all other strains? Which are the genes that are uniquely DE by SS8? 
```{r, fig.align='center'}
## CONTRAST COMPARING WT10 to the other host types
cont.matrix2_host <- makeContrasts(SS8.vs.WT10 = StrainStrain8 - StrainStrain10, SS3.vs.WT10 = StrainStrain3 - StrainStrain10, SS5.vs.WT10 = StrainStrain5 - StrainStrain10, levels=design)
cont.matrix2_host

## fit contrasts
fit.cont2_host <- contrasts.fit(gene_host_fit_NoIntercept, cont.matrix2_host) %>% eBayes()

genes_sign_WT10_Intercept_host <- decideTests(fit.cont2_host) %>% as.data.frame() %>%  bind_cols(fit.cont2_host$genes)

vennDiagram(genes_sign_WT10_Intercept_host[,1:3],include=c("up","down"),
            counts.col=c("brown3", "darkblue"),
            circle.col = c("darkred", "blue", "green4"))
title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)


#genes_sign_WT10_Intercept_host2 <- decideTests(fit.cont2_host, lfc=1) %>% as.data.frame() %>%  bind_cols(fit.cont2_host$genes)
#vennDiagram(genes_sign_WT10_Intercept_host2[,1:3],include=c("up","down"),
#            counts.col=c("brown3", "darkblue"),
#            circle.col = c("darkred", "blue", "green4"))
#title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)


```

<br><br>


## Exploratory analysis 1
### Heat map of general gene expression.
```{r DE-heatmap-plot-host, fig.align='center'}
# Filter genes called as DE to control in any strain

any_DE_genes_host <- decideTests(fit.cont1_host, lfc=1) %>% 
  as.data.frame() %>% 
  bind_cols(fit.cont2_host$genes) %>% 
  filter_if(~ is.numeric(.), any_vars(. != 0))
#
# Separate out ones called as DE in *all* strains as well
all_DE_genes_host <- any_DE_genes_host%>% 
  filter_if(~ is.numeric(.), all_vars(. != 0))
#
## Heatmap of DE genes
# Normalise expression per-gene
normalised_DE_gene_expression_host <- gene_host_v[gene_host_v$genes$gene %in% all_DE_genes_host$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(host_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup
#
# Cluster to show patterns
clustering_host <- normalised_DE_gene_expression_host %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()
#
# Plot it
expression_heatmap_host_top <- normalised_DE_gene_expression_host %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_host$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_text(size=8)) + # With Text
  scale_x_discrete(labels = host_samples$Strain %>% str_remove("Strain")) +
  xlab("Sample Strain")
#
expression_heatmap_host_top
#


#Can also include the actual DE calls
DE_calls_host <- any_DE_genes_host %>% 
  gather(Strain, DE_call, -gene, -Annotation) %>% 
  mutate(
    Strain = str_remove(Strain, "StrainStrain"), 
    gene = factor(gene, levels = gene %>% unique %>% .[clustering_host$order] )
    ) %>% 
  ggplot(aes(x = Strain, y = gene, fill = as.factor(DE_call))) + 
  geom_tile() + 
  theme_minimal() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(), legend.position = "none") +
  scale_x_discrete() +
  xlab("DE Calls") + 
  scale_fill_manual(values = c("blue", "white","red"))
#
cowplot::plot_grid(DE_calls_host, expression_heatmap_host_top + ylab(NULL), align = "v", rel_widths = c(1,5), axis = "tb")

```
<br><br>



## Exploratory analysis 2
### Hypothesis testing
<br>
Filter genes according to annotation. Which are heat stress related genes that are differentially expressed between the wild type and selected strains?
```{r, fig.align='center'}
## Filter based on gene host_annotations

# Heat stress associated genes
fit.cont1host_heat <- topTable(fit.cont1_host, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[H|h]eat|[S|s]uperoxide|[G|g]lutathione|[C|c]haperone|[P|p]eroxidase|[H|h]sp"))
fit.cont1host_heat

# Heat map of Top significant genes
fit.cont1host_TopSign <- topTable(fit.cont1_host, number = 500)



# Heat map of Top significant unique genes
SS8_unique_up_host <- filter(genes_sign_WT10_Intercept_host, SS3.vs.WT10 == 0 | SS3.vs.WT10 == -1, SS5.vs.WT10 == 0 | SS5.vs.WT10 == -1, SS8.vs.WT10 == 1) %>% .[order(.$Annotation),]
fit.cont1_host_SS8unique_up <- topTable(fit.cont1_host, number = Inf, p.value = 0.05) %>% filter(gene %in% SS8_unique_up_host$gene) # %>% str_detect(SS_8_unique_up$gene))
SS8_unique_up_host_TopSign <- head(fit.cont1_host_SS8unique_up, 15)
#
SS8_unique_down_host <- filter(genes_sign_WT10_Intercept_host, SS3.vs.WT10 == 0 | SS3.vs.WT10 == 1, SS5.vs.WT10 == 0 | SS5.vs.WT10 == 1, SS8.vs.WT10 == -1) %>% .[order(.$Annotation),]
fit.cont1_host_SS8unique_down <- topTable(fit.cont1_host, number = Inf, p.value = 0.05) %>% filter(gene %in% SS8_unique_down_host$gene) # %>% str_detect(SS_8_unique_up$gene))
SS8_unique_down_host_TopSign <- head(fit.cont1_host_SS8unique_down, 15)
#
fit.cont1_host_SS8unique_TopSign <- rbind(SS8_unique_up_host_TopSign, SS8_unique_down_host_TopSign)


# Heat map of Top significant common genes
SS3_5_8_common_up_host <- filter(genes_sign_WT10_Intercept_host, SS3.vs.WT10 == 1, SS8.vs.WT10 == 1, SS8.vs.WT10 == 1) %>% .[order(.$Annotation),]
fit.cont2_host_SS358common_up <- topTable(fit.cont2_host, number = Inf, p.value = 0.05) %>% filter(gene %in% SS3_5_8_common_up_host$gene) # %>% str_detect(SS_8_unique_up$gene))
SS358_common_up_host_TopSign <- head(fit.cont2_host_SS358common_up, 25)
#
SS3_5_8_common_down_host <- filter(genes_sign_WT10_Intercept_host, SS3.vs.WT10 == -1, SS8.vs.WT10 == -1, SS8.vs.WT10 == -1) %>% .[order(.$Annotation),]
fit.cont2_host_SS358common_down <- topTable(fit.cont2_host, number = Inf, p.value = 0.05) %>% filter(gene %in% SS3_5_8_common_down_host$gene) # %>% str_detect(SS_8_unique_up$gene))
SS358_common_down_host_TopSign <- head(fit.cont2_host_SS358common_down, 25)
#
fit.cont2_host_SS358common_TopSign <- rbind(SS358_common_down_host_TopSign, SS358_common_up_host_TopSign)





####
# Data feed for heat map, change as required..
heatmap.data.host <- fit.cont1host_heat
#

normalised_DE_gene_expression_host <- gene_host_v[gene_host_v$genes$gene %in% heatmap.data.host$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(host_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup
#
# Cluster to show patterns
clustering_host <- normalised_DE_gene_expression_host %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()
#
# Plot it (remove "#" theme to remove x axis labels)
expression_heatmap_host_Topsign <- normalised_DE_gene_expression_host %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_host$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  scale_x_discrete(labels = host_samples$Strain %>% str_remove("Strain")) +
  scale_y_discrete(labels = normalised_DE_gene_expression_host %>% group_by(gene, Annotation) %>% summarise() %>% arrange(clustering_host$order) %>% .$Annotation)
  xlab("Sample Strain")
#
## Show heat map
expression_heatmap_host_Topsign



```
<br><br>
  

  
## Plotting, plot by gene and by exon.
```{r, fig.align='center'}

## Plot by gene, according to annotation
plot_host_gene <- function(gene) {
  #expn <- gene_host_v$E[gene_host_v$genes$Annotation == gene, ]
  expn <- gene_host_v$E[gene_host_v$genes$gene == gene, ]
  gene_host_v$targets %>% 
    mutate(expn = expn) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    theme(legend.position = "none")
}

# gene of choice
par(mar=c(5,6,3,1))
layout(matrix(c(1,2), 2, 2, byrow = TRUE))
#plot_host_gene("aten_0.1.m1.446.m1")
#plot_host_gene("aten_0.1.m1.4452.m1")
plot_host_gene("aten_0.1.m1.24908.m1")
plot_host_gene("aten_0.1.m1.31125.m1")
#plot_host_gene("aten_0.1.m1.17022.m1")
plot_host_gene("aten_0.1.m1.10027.m1")

dev.off() 




## Plot by exon, according to annotation
plot_host_gene_by_exon <- function(gene) {
  host_v$E %>% 
    as.data.frame() %>% 
    filter(str_detect(host_v$genes$EntrezGeneID, gene)) %>% 
    mutate(exon = host_v$genes$EntrezGeneID[str_detect(host_v$genes$EntrezGeneID, gene)]) %>% 
    gather(sample, expn, -exon) %>% 
    left_join(host_v$targets %>% as.data.frame() %>% rownames_to_column("sample")) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    facet_wrap(~factor(exon, levels = stringr::str_sort(unique(exon), numeric = T))) +
    theme(legend.position = "none")
}


# exons of gene of choice
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
#plot_host_gene_by_exon("aten_0.1.m1.446.m1")
#plot_host_gene_by_exon("aten_0.1.m1.4452.m1")
plot_host_gene_by_exon("aten_0.1.m1.24908.m1")
#plot_host_gene_by_exon("aten_0.1.m1.17022.m1")


```

***
<br><br><br>



## Data export host
code silenced
```{r, eval=FALSE}
# Specific pairwise comparisons. What separates SS8 from respective strains?
library(openxlsx)
SampleGroup<-"host_WT10_intercept_VS_pairwise"
Title <- paste(SampleGroup,"DGE-Table")
FileName <- paste(Title,".xlsx")
limma.resA <- topTable(fit.cont2_host,coef="SS8.vs.WT10",sort.by="p",n="Inf")
limma.resB <- topTable(fit.cont2_host,coef="SS3.vs.WT10",sort.by="p",n="Inf")
limma.resC <- topTable(fit.cont2_host,coef="SS5.vs.WT10",sort.by="p",n="Inf")
list_of_datasets_host <- list("SS8.vs.WT10" = limma.resA, "SS3.vs.WT10" = limma.resB, "SS5.vs.WT10" = limma.resC)
write.xlsx(list_of_datasets_host, file=FileName, sep="\t", row.names=TRUE)
#

# General adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
host_WT10_intercept_VS_all <- topTable(fit.cont2_host, n=Inf) %>% write.table(, file="host_WT10_intercept_VS_all_topTable.csv", sep="\t", quote=F)
##
# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN GOSeq and REVIGO
write.xlsx(genes_sign_WT10_Intercept_host, file="host_SS8_unique_DGE_expression_list_WT10.xlsx", sheetName="SS_8_vs_WT10_unique", append=TRUE, row.names=FALSE)
##
# 


# Export Toptable unique genes
fit.cont1_host_SS8unique_TopSign <- rbind(fit.cont1_host_SS8unique_up, fit.cont1_host_SS8unique_down)
library(openxlsx)
SampleGroup<-"fit.cont1_host_SS8unique_TopSign"
Title <- paste(SampleGroup,"SS8")
FileName <- paste(Title,".xlsx")
write.xlsx(fit.cont1_host_SS8unique_TopSign, file=FileName, sep="\t", row.names=TRUE)

# Export Toptable common genes
fit.cont2_host_SS358common_TopSign <- rbind(fit.cont2_host_SS358common_up, fit.cont2_host_SS358common_down)
library(openxlsx)
SampleGroup<-"fit.cont2_host_SS358common_TopSign"
Title <- paste(SampleGroup,"SS8")
FileName <- paste(Title,".xlsx")
write.xlsx(fit.cont2_host_SS358common_TopSign, file=FileName, sep="\t", row.names=TRUE)


```
 
<br>
<br>
<br>
  



# __2 Symbiont expression__
## Loading data

Read in the symbiont count matrix and convert to log2cpm
``` {r symbiont-data, message=FALSE, warning=FALSE, fig.align='center'}
symbiont_counts <- read_tsv("data/MATRIX_Counts-FINAL-Symbiont.txt")
symbiont_samples <- read_tsv("data/SampleInfo_Symbiont-STAR.txt")
symbiont_annotations <- read_tsv("data/Symbiont_diamond_annot_20190117_1327.txt")
#symbiont_exon_annnotations <- read_tsv("data/Symbiont-Gene-Annotation.txt")
```

```{r}
# Exons
symbiont_DGE <- DGEList(symbiont_counts[,-1], genes = symbiont_counts[,1], samples = symbiont_samples)

filt <- filterByExpr(symbiont_DGE, design = model.matrix(~Strain, data = symbiont_DGE$samples), min.count = 40)

symbiont_filtered1 <- symbiont_DGE[filt, , keep.lib.sizes = F]

symbiont_filtered2 <- calcNormFactors(symbiont_filtered1)

symbiont_v <- voom(symbiont_filtered2, design = model.matrix(~Strain, data = symbiont_filtered2$samples), plot = F)

```
<br><br>

## Collate to gene level
Might be better for some uses to look at whole gene level expression initally before drilling down to exon level data.
```{r symbiont-gene-level-counts, fig.align='center'}
gene_symbiont_counts <- symbiont_counts %>% 
  separate(EntrezGeneID, c("gene", "exon"), sep="\\.exon") %>% 
  group_by(gene) %>% summarise_if(is.numeric, sum)

# check that symbiont_annotations and genes IDs match.. needs to be all TRUE events
table((symbiont_annotations[,1] == (gene_symbiont_counts[,1])))

# then bind symbiont_annotations to gene list
gene_symbiont_DGE <- DGEList(gene_symbiont_counts[,-1], genes = gene_symbiont_counts[,1]%>% bind_cols(symbiont_annotations[,-1]), samples = symbiont_samples)

# check that annotation table and gene_symbiont_DGE match.. needs to be all TRUE events. Two columns to match
table((gene_symbiont_DGE$genes == (symbiont_annotations)))

gene_symbiont_filt <- filterByExpr(gene_symbiont_DGE, design = model.matrix(~0 + Strain, data = gene_symbiont_DGE$samples), min.count = 40)

gene_symbiont_filtered1 <- gene_symbiont_DGE[gene_symbiont_filt, keep.lib.sizes = F]


gene_symbiont_filtered2 <- calcNormFactors(gene_symbiont_filtered1)


# Check filter cut off before and after
par(mfrow=c(2,2))
mean_log_cpm3 <- aveLogCPM(gene_symbiont_DGE$counts)
filter_threshold <- 0.5
hist(mean_log_cpm3, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm3); abline(h=filter_threshold)
#
mean_log_cpm4 <- aveLogCPM(gene_symbiont_filtered1$counts)
filter_threshold <- 0.5
hist(mean_log_cpm4, breaks=200, ylim=c(0,2500)); abline(v=filter_threshold)
qqnorm(mean_log_cpm4); abline(h=filter_threshold)

# Check normalised and unnormalised data
par(mfrow=c(1,2))
Symbiont_lcpm1 <- cpm(gene_symbiont_filtered1, log=TRUE)
boxplot(Symbiont_lcpm1, las=2, main="")
title(main="A. Example: Unnormalised data",ylab="Log-cpm")
#
gene_symbiont_filtered2$samples$norm.factors
Symbiont_lcpm2 <- cpm(gene_symbiont_filtered2, log=TRUE)
boxplot(Symbiont_lcpm2, las=2, main="")
title(main="B. Example: Normalised data",ylab="Log-cpm")




# check that data is still all matching up
table(symbiont_samples$SampleName == colnames(gene_symbiont_filtered2))

gene_symbiont_v <- voom(gene_symbiont_filtered2, design = model.matrix(~ 0 + Strain, data = gene_symbiont_filtered2$samples), plot = T)

#
gene_symbiont_fit_NoIntercept <- lmFit(gene_symbiont_v, design = model.matrix(~0 +Strain, data = gene_symbiont_filtered2$samples)) %>% eBayes()
#
plotSA(gene_symbiont_fit_NoIntercept)

design <- model.matrix(~0 +Strain, data = gene_symbiont_filtered2$samples)
#
colnames(design)

```
Fig. 4 Limma voom model.
<br><br>
  
## MDS to check overall expression differences
```{r, fig.align='center'}
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
plotMDS(symbiont_v, main = "exon counts", labels = symbiont_v$targets$FileName %>% str_remove("symbiont\\."), col = factor(symbiont_v$targets$Strain %>% str_replace("symbiont\\.(.*?)\\..*", "\\1")) %>% as.integer())
#
plotMDS(gene_symbiont_v, main = "gene counts",labels = gene_symbiont_v$targets$FileName %>% str_remove("symbiont\\."), col = factor(gene_symbiont_v$targets$Strain %>% str_replace("symbiont\\.(.*?)\\..*", "\\1")) %>% as.integer())

```
Fig. 5 MDS plots.
<br><br>  


## Contrasts 1
<br>
```{r Make contrasts - Symbiont, fig.align='center'}

## CONTRASTS
## CONTRAST COMPARING THE SS8 to all other symbiont types. 
## What are the general adaptations of SS8?
cont.matrix_sym <- makeContrasts(SS8.vs.WT10 = StrainStrain8 - StrainStrain10, SS8.vs.SS3 = StrainStrain8 - StrainStrain3, SS8.vs.SS5 = StrainStrain8 - StrainStrain5, levels=design)
cont.matrix_sym

## fit contrasts
fit.cont1_symbiont <- contrasts.fit(gene_symbiont_fit_NoIntercept, cont.matrix_sym) %>% eBayes()
# fit.cont
# dim(fit.cont1symbiont)
summary(decideTests(fit.cont1_symbiont))

#view fit stats
topTable(fit.cont1_symbiont)

genes_sign_SS8_Intercept_symbiont <- decideTests(fit.cont1_symbiont) %>% as.data.frame() %>%  bind_cols(fit.cont1_symbiont$genes)

vennDiagram(genes_sign_SS8_Intercept_symbiont[,1:3],include=c("up","down"),
            counts.col=c("brown3", "darkblue"),
            circle.col = c("darkred", "blue", "green4"))
title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)

```

Fig. 6 Venndiagram showing genes that are significantly different expressed between the SS8 and respective other strains. The differentially expressed genes show the general difference of SS8 to all other strains, i.e. the specific adaptations of SS8.

<br><br>

DATA EXPORT at the end of the script..

<br><br>


## Contrasts 2
<br>
What makes SS8 unique compared to all other strains? Which are the genes that are uniquely DE by SS8? 
```{r, fig.align='center'}
## CONTRAST COMPARING THE SS8 to the other symbiont types, here WT10
cont.matrix2_symbiont <- makeContrasts(SS8.vs.WT10 =  StrainStrain8 - StrainStrain10, SS3.vs.WT10 =  StrainStrain3 - StrainStrain10, SS5.vs.WT10 = StrainStrain5 - StrainStrain10, levels=design)
cont.matrix2_symbiont

## fit contrasts
fit.cont2_symbiont <- contrasts.fit(gene_symbiont_fit_NoIntercept, cont.matrix2_symbiont) %>% eBayes()

genes_sign_WT10_Intercept_symbiont <- decideTests(fit.cont2_symbiont) %>% as.data.frame() %>%  bind_cols(fit.cont2_symbiont$genes)
vennDiagram(genes_sign_WT10_Intercept_symbiont[,1:3],include=c("up","down"),
            counts.col=c("brown3", "darkblue"),
            circle.col = c("darkred", "blue", "green4"))
title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)


#genes_sign_WT10_Intercept_symbiont2 <- decideTests(fit.cont2_symbiont, lfc=0.5) %>% as.data.frame() %>%  bind_cols(fit.cont2_symbiont$genes)
#vennDiagram(genes_sign_WT10_Intercept_symbiont2[,1:3],include=c("up","down"),
#            counts.col=c("brown3", "darkblue"),
#            circle.col = c("darkred", "blue", "green4"))
#title("Venndiagram symbiont. SS8 strain vs all other strains [all genes]", line = 1)

```

<br><br>


## Exploratory analysis 1
### Heat map of general gene expression.
```{r DE-heatmap-plot-symbiont, fig.align='center'}
# Filter genes called as DE to control in any strain

# any_DE_genes_symbiont <- decideTests(fit.cont1_symbiont, lfc=0.5) %>% 
any_DE_genes_symbiont <- decideTests(fit.cont1_symbiont) %>% 
  as.data.frame() %>% 
  bind_cols(fit.cont2_symbiont$genes) %>% 
  filter_if(~ is.numeric(.), any_vars(. != 0))
#
# Separate out ones called as DE in *all* strains as well
all_DE_genes_symbiont <- any_DE_genes_symbiont%>% 
  filter_if(~ is.numeric(.), all_vars(. != 0))
#
## Heatmap of DE genes
# Normalise expression per-gene
normalised_DE_gene_expression_symbiont <- gene_symbiont_v[gene_symbiont_v$genes$gene %in% all_DE_genes_symbiont$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(symbiont_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup
#
# Cluster to show patterns
clustering_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()
#
# Plot it
expression_heatmap_symbiont_top <- normalised_DE_gene_expression_symbiont %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_symbiont$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  #theme(axis.ticks.y = element_blank(), axis.text.y = element_text(size=8)) + #Text
  scale_x_discrete(labels = symbiont_samples$Strain %>% str_remove("Strain")) +
  xlab("Sample Strain")
#
expression_heatmap_symbiont_top


#Can also include the actual DE calls
DE_calls_symbiont <- any_DE_genes_symbiont %>% 
  gather(Strain, DE_call, -gene, -Annotation) %>% 
  mutate(
    Strain = str_remove(Strain, "StrainStrain"), 
    gene = factor(gene, levels = gene %>% unique %>% .[clustering_symbiont$order] )
    ) %>% 
  ggplot(aes(x = Strain, y = gene, fill = as.factor(DE_call))) + 
  geom_tile() + 
  theme_minimal() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(), legend.position = "none") +
  scale_x_discrete() +
  xlab("DE Calls") + 
  scale_fill_manual(values = c("blue", "white","red"))
#
cowplot::plot_grid(DE_calls_symbiont, expression_heatmap_symbiont_top + ylab(NULL), align = "v", rel_widths = c(1,5), axis = "tb")


```
<br><br>



## Exploratory analysis 2
### Hypothesis testing
<br>
Filter genes according to annotation. Which are heat stress related genes that are differentially expressed between the wild type and selected strains?
```{r, fig.align='center'}
## Filter based on gene symbiont_annotations

# Heat stress associated genes
fit.cont1symbiont_heat <- topTable(fit.cont1_symbiont, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[H|h]eat|[S|s]uperoxide|[G|g]lutathione|[C|c]haperone|[P|p]eroxidase|[H|h]sp|[D|d]naj"))
fit.cont1symbiont_heat

# Calvin-Cycle associated genes
fit.cont2symbiont_calvin <- topTable(fit.cont1_symbiont, number = Inf, p.value = 0.05) %>% filter(Annotation %>% str_detect("[R|r]ibulose|[R|r]bcL|[P|p]hosphoglycolate phosphatase|[P|p]hosphoglycerate phosphatase| [P|p]hosphoglycerate kinase|[F|f]ructose-bisphosphate"))
fit.cont2symbiont_calvin

# Heat map of Top significant genes
fit.cont1symbiont_TopSign <- topTable(fit.cont2_symbiont, number = 50)
fit.cont1symbiont_TopSign

# Heat map of Top significant unique genes
SS8_unique_up_symbiont <- filter(genes_sign_WT10_Intercept_symbiont, SS3.vs.WT10 == 0 | SS3.vs.WT10 == -1, SS5.vs.WT10 == 0 | SS5.vs.WT10 == -1, SS8.vs.WT10 == 1) %>% .[order(.$Annotation),]
fit.cont1_symbiont_SS8unique_up <- topTable(fit.cont1_symbiont, number = Inf, p.value = 0.05) %>% filter(gene %in% SS8_unique_up_symbiont$gene) # %>% str_detect(SS_8_unique_up$gene))
SS8_unique_up_symbiont_TopSign <- head(fit.cont1_symbiont_SS8unique_up, 15)
#
SS8_unique_down_symbiont <- filter(genes_sign_WT10_Intercept_symbiont, SS3.vs.WT10 == 0 | SS3.vs.WT10 == 1, SS5.vs.WT10 == 0 | SS5.vs.WT10 == 1, SS8.vs.WT10 == -1) %>% .[order(.$Annotation),]
fit.cont1_symbiont_SS8unique_down <- topTable(fit.cont1_symbiont, number = Inf, p.value = 0.05) %>% filter(gene %in% SS8_unique_down_symbiont$gene) # %>% str_detect(SS_8_unique_up$gene))
SS8_unique_down_symbiont_TopSign <- head(fit.cont1_symbiont_SS8unique_down, 15)
#
fit.cont1_symbiont_SS8unique_TopSign <- rbind(SS8_unique_up_symbiont_TopSign, SS8_unique_down_symbiont_TopSign)

# Heat map of Top significant common genes
SS3_5_8_common_up_symbiont <- filter(genes_sign_WT10_Intercept_symbiont, SS3.vs.WT10 == 1, SS8.vs.WT10 == 1, SS8.vs.WT10 == 1) %>% .[order(.$Annotation),]
fit.cont2_symbiont_SS358common_up <- topTable(fit.cont2_symbiont, number = Inf, p.value = 0.05) %>% filter(gene %in% SS3_5_8_common_up_symbiont$gene) # %>% str_detect(SS_8_unique_up$gene))
SS358_common_up_symbiont_TopSign <- head(fit.cont2_symbiont_SS358common_up, 25)
#
SS3_5_8_common_down_symbiont <- filter(genes_sign_WT10_Intercept_symbiont, SS3.vs.WT10 == -1, SS8.vs.WT10 == -1, SS8.vs.WT10 == -1) %>% .[order(.$Annotation),]
fit.cont2_symbiont_SS358common_down <- topTable(fit.cont2_symbiont, number = Inf, p.value = 0.05) %>% filter(gene %in% SS3_5_8_common_down_symbiont$gene) # %>% str_detect(SS_8_unique_up$gene))
SS358_common_down_symbiont_TopSign <- head(fit.cont2_symbiont_SS358common_down, 25)
#
fit.cont2_symbiont_SS358common_TopSign <- rbind(SS358_common_down_symbiont_TopSign, SS358_common_up_symbiont_TopSign)


####
# Data feed for heat map, change as required..
heatmap.data.symbiont <- fit.cont2symbiont_calvin

#
normalised_DE_gene_expression_symbiont <- gene_symbiont_v[gene_symbiont_v$genes$gene %in% heatmap.data.symbiont$gene, ] %>% 
  as.data.frame() %>% 
  gather(SampleName, log2cpm, -gene, -Annotation) %>% 
  left_join(symbiont_samples) %>% 
  group_by(gene) %>%
  mutate(scaled_log2cpm = scale(log2cpm)) %>% 
  ungroup
#
# Cluster to show patterns
clustering_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  dplyr::select(gene, SampleName, scaled_log2cpm) %>% 
  spread(SampleName, scaled_log2cpm) %>% 
  dplyr::select(-gene) %>% 
  cluster::agnes()
#
# Plot it
expression_heatmap_symbiont <- normalised_DE_gene_expression_symbiont %>% 
  mutate(gene = factor(gene, levels = gene %>% unique %>% .[clustering_symbiont$order] )) %>% 
  ggplot(aes(x=SampleName, y = gene, fill = scaled_log2cpm)) +
  geom_tile() +
  scale_fill_gradient2(low = "#0073ff", high = "#ff8c00") +
  theme_bw() +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank()) +
  scale_x_discrete(labels = symbiont_samples$Strain %>% str_remove("Strain")) +
  scale_y_discrete(labels = normalised_DE_gene_expression_symbiont %>% group_by(gene, Annotation) %>% summarise() %>% arrange(clustering_symbiont$order) %>% .$Annotation)
  xlab("Sample Strain")
#
#
# Show heat map
expression_heatmap_symbiont


```
<br><br>
  

## Plotting, plot by gene and by exon.
```{r, fig.align='center'}

## Plot by gene, according to annotation
plot_symbiont_gene <- function(gene) {
  #expn <- gene_symbiont_v$E[gene_symbiont_v$genes$Annotation == gene, ]
  expn <- gene_symbiont_v$E[gene_symbiont_v$genes$gene == gene, ]
  gene_symbiont_v$targets %>% 
    mutate(expn = expn) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    theme(legend.position = "none")
}

# gene of choice
par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))

plot_symbiont_gene("SymbC1.scaffold7634.1.m1")  # Superoxide dismutase, SS8 UNIQUE
#plot_symbiont_gene("SymbC1.scaffold10965.1.m1") # family chaperone, SS8 UNIQUE
#plot_symbiont_gene("SymbC1.scaffold213.10.m1")  # oxidoreductase activity, SS8 UNIQUE
#plot_symbiont_gene("SymbC1.scaffold2883.4.m1")  # Lactoylglutathione lyase
#plot_symbiont_gene("SymbC1.scaffold714.1.m1")   # Chaperone chloroplastic
#plot_symbiont_gene("SymbC1.scaffold1286.6.m1")  #	Glutathione synthetase
#plot_symbiont_gene("SymbC1.scaffold2650.2.m1")  #	E3 ubiquitin- ligase HERC1
#plot_symbiont_gene("SymbC1.scaffold12561.1.m1") #	Cytochrome mitochondrial

plot_symbiont_gene("SymbC1.scaffold9320.1.m1")  # Heat shock 70 kDa 

plot_symbiont_gene("SymbC1.scaffold6940.1.m1")  # ribulose 1,5-bisphosphate carboxylase oxygenase form
plot_symbiont_gene("SymbC1.scaffold6940.2.m1")  # ribulose 1,5-bisphosphate carboxylase oxygenase form
plot_symbiont_gene("SymbC1.scaffold9010.1.m1")  # ribulose bisphosphate chloroplastic
plot_symbiont_gene("SymbC1.scaffold23876.1.m1") # ribulose bisphosphate carboxylase	

#plot_symbiont_gene("SymbC1.scaffold1364.11")   # High affinity nitrate transporter
#plot_symbiont_gene("SymbC1.scaffold912.8.m1")  # phosphoglycolate phosphatase
#plot_symbiont_gene("SymbC1.scaffold9459.1")  # bifunctional aconitate hydratase 2 2-methylisocitrate dehydratase

## Plot by exon, according to annotation
plot_symbiont_gene_by_exon <- function(gene) {
  symbiont_v$E %>% 
    as.data.frame() %>% 
    filter(str_detect(symbiont_v$genes$EntrezGeneID, gene)) %>% 
    mutate(exon = symbiont_v$genes$EntrezGeneID[str_detect(symbiont_v$genes$EntrezGeneID, gene)]) %>% 
    gather(sample, expn, -exon) %>% 
    left_join(symbiont_v$targets %>% as.data.frame() %>% rownames_to_column("sample")) %>% 
    ggplot(aes(x=Strain, y = expn, colour = CellType)) + 
    geom_point(position = position_jitter(width = 0.2)) +
    stat_summary(colour = "black") +
    ggtitle(gene) +
    facet_wrap(~factor(exon, levels = stringr::str_sort(unique(exon), numeric = T))) +
    theme(legend.position = "none")
}


# exons of gene of choice
#par(mar=c(5,6,3,1))
#layout(matrix(c(1,2), 2, 2, byrow = TRUE))
plot_symbiont_gene_by_exon("SymbC1.scaffold7634.1.m1")    # Superoxide dismutase, SS8 UNIQUE
#plot_symbiont_gene_by_exon("SymbC1.scaffold10965.1.m1")   # family chaperone, SS8 UNIQUE UNIQUE
#plot_symbiont_gene_by_exon("SymbC1.scaffold213.10.m1")    # oxidoreductase activity, SS8 UNIQUE
#plot_symbiont_gene_by_exon("SymbC1.scaffold2883.4.m1")    # Lactoylglutathione lyase
#plot_symbiont_gene_by_exon("SymbC1.scaffold714.1.m1")     # Chaperone chloroplastic
plot_symbiont_gene_by_exon("SymbC1.scaffold1787.3.m1")    # hsp organising protein



```

***
<br><br><br>



## Data export symbiont
code silenced
```{r, eval=FALSE}

# Specific pairwise comparisons. What separates WT10 from respective strains?
library(openxlsx)
SampleGroup<-"symbiont_WT10_intercept_VS_pairwise"
Title <- paste(SampleGroup,"DGE-Table_NEW")
FileName <- paste(Title,".xlsx")
limma.resA <- topTable(fit.cont2_symbiont,coef="SS8.vs.WT10",sort.by="p",n="Inf")
limma.resB <- topTable(fit.cont2_symbiont,coef="SS3.vs.WT10",sort.by="p",n="Inf")
limma.resC <- topTable(fit.cont2_symbiont,coef="SS5.vs.WT10",sort.by="p",n="Inf")
list_of_datasets_symbiont <- list("SS8.vs.WT10" = limma.resA, "SS3.vs.WT10" = limma.resB, "SS5.vs.WT10" = limma.resC)
write.xlsx(list_of_datasets_symbiont, file=FileName, sep="\t", row.names=TRUE)
#
# Gerneral adaptations of respective strains. Take this table for GO analysis with MWU/Fisher script. 
symbiont_WT10_intercept_VS_all <- topTable(fit.cont2_symbiont, n=Inf) %>% write.table(, file="symbiont_WT10_intercept_VS_all_topTable_NEW.csv", sep="\t", quote=F)
##

# USE THESE FOR INDIVIDUAL GENE COMPARISONS IN GOSeq and REVIGO
write.xlsx(genes_sign_WT10_Intercept_symbiont, file="symbiont_unique_DGE_expression_list_NEW.xlsx", sheetName="SS8_vs_WT10_unique", append=TRUE, row.names=FALSE)
#


# Export Toptable unique genes
fit.cont1_symbiont_SS8unique_TopSign <- rbind(fit.cont1_symbiont_SS8unique_up, fit.cont1_symbiont_SS8unique_down)
library(openxlsx)
SampleGroup<-"fit.cont1_symbiont_SS8unique_TopSign"
Title <- paste(SampleGroup,"SS8")
FileName <- paste(Title,".xlsx")
write.xlsx(fit.cont1_symbiont_SS8unique_TopSign, file=FileName, sep="\t", row.names=TRUE)

# Export Toptable common genes
fit.cont2_symbiont_SS358common_TopSign <- rbind(fit.cont2_symbiont_SS358common_up, fit.cont2_symbiont_SS358common_down)
library(openxlsx)
SampleGroup<-"fit.cont2_symbiont_SS358common_TopSign"
Title <- paste(SampleGroup,"SS8")
FileName <- paste(Title,".xlsx")
write.xlsx(fit.cont2_symbiont_SS358common_TopSign, file=FileName, sep="\t", row.names=TRUE)


```

***
<br><br>



# __3 Additional Code__
## MDS, statistics with adonis - Host
```{r, fig.align='center', warning = FALSE, message = FALSE}
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)

#
######## MDS-WITH ADONIS
# DIFFERENT INPUT - USE logCPM on counts before voom
#head(gene_host_filtered1$counts)
#head(logcpmtmm)
#logcounts <- gene_host_filtered1$counts
  
#head(gene_host_filtered2$counts)
#logcounts <- gene_host_filtered2$counts

myCPM_host <- cpm(gene_host_filtered2$counts, log = TRUE)
#head(myCPM_host)
logcounts_host <- myCPM_host

#head(gene_host_v$E)
#logcounts <- gene_host_v$E
#head(logcounts)


#
# Make Factors
CellType=as.character(host_samples$CellType)
Strain=as.character(host_samples$Strain)
#CellType
#Strain
conditions=data.frame(cbind(CellType,Strain))
#conditions
#
#

# Calculate distances etc..
dds.pcoa3_host=pcoa(vegdist(t(logcounts_host),method="euclidean")/1000)
# export values
scores3_host=dds.pcoa3_host$vectors

#
#head(scores3)
#
#
SampleGroup<-"Host-STAR"
#SampleGroup<-"Symb-STAR"
time<-format(Sys.time(),"%d-%m-%Y")
#
#Title <- paste(SampleGroup,"MDS-logcounts-euclidean")
#FileName <- paste(Title,time,".jpg")
#jpeg(file=FileName, width = 2020, height = 1080, res = 200)

#par(mfrow=c(1,2))
#

plot(scores3_host[,1], scores3_host[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","orange","orange","orange","orange","orange","Blue","Blue","Blue","Blue","Blue","Blue"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3_host,CellType,label=T)
ordiellipse(scores3_host,CellType)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)

plot(scores3_host[,1], scores3_host[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","orange","orange","orange","orange","orange","Blue","Blue","Blue","Blue","Blue","Blue"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3_host,Strain,label=T)
ordiellipse(scores3_host,Strain)
#
Title <- paste(SampleGroup,"MDS-Plot")
#
title(Title)

dev.off() 

## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(CellType)),type="s",radius=0.005)

##
## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(Strain)),type="s",radius=0.005)

##
##
##

## STATISTIC TESTS - Adonis
##
Title <- paste(SampleGroup,"Adonis")
FileName <- paste("5-",Title,time,".txt")

ad_host=adonis(t(logcounts_host)~CellType+Strain, permutations = 999, data=conditions, method="euclidean")
ad_host 
#
#capture.output(ad_host,file=FileName)
#
Title <- paste(SampleGroup,"TukeyHSD")
FileName <- paste("5-",Title,time,".txt")
ad_host=adonis(t(logcounts_host)~Strain,data=conditions,method="euclidean")
ad_host
pco1=scores3_host[,1]
TukeyHSD(aov(pco1~Strain))
#
#capture.output(TukeyHSD(aov(pco1~Strain)),file=FileName)

```

<br><br>




## MDS, statistics with adonis - Symbiont
```{r, fig.align='center'}

library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)

#
######## MDS-WITH ADONIS
#
## MDS PLOT FROM ANY MATRIX

# DIFFERENT INPUT - USE logCPM on counts before voom
#head(gene_symbiont_filtered1$counts)
#head(logcpmtmm)
#logcounts <- gene_symbiont_filtered1$counts
  
#head(gene_symbiont_filtered2$counts)
#logcounts <- gene_symbiont_filtered2$counts

myCPM_symbiont <- cpm(gene_symbiont_filtered2$counts, log = TRUE)
#head(myCPM_symbiont)
logcounts_symbiont <- myCPM_symbiont

#head(gene_symbiont_v$E)
#logcounts <- gene_symbiont_v$E
#head(logcounts)


#
# Make Factors
CellType=as.character(symbiont_samples$CellType)
Strain=as.character(symbiont_samples$Strain)
#CellType
#Strain
conditions=data.frame(cbind(CellType,Strain))
#conditions
#

# Calculate distances etc..
dds.pcoa3_symbiont=pcoa(vegdist(t(logcounts_symbiont),method="euclidean")/1000)
# export values
scores3_symbiont=dds.pcoa3_symbiont$vectors

#
# head(scores3)
#
#
#SampleGroup<-"Host-STAR"
SampleGroup<-"Symb-STAR"
time<-format(Sys.time(),"%d-%m-%Y")
#
#Title <- paste(SampleGroup,"MDS-logcounts-euclidean-logCPM")
#FileName <- paste(Title,time,".jpg")
#jpeg(file=FileName, width = 2020, height = 1080, res = 200)
#par(mfrow=c(1,2))
#
plot(scores3_symbiont[,1], scores3_symbiont[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","orange","orange","orange","orange","orange","Blue","Blue","Blue","Blue","Blue","Blue"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3_symbiont,CellType,label=T)
ordiellipse(scores3_symbiont,CellType)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)
#
plot(scores3_symbiont[,1], scores3_symbiont[,2], col="Black", bg=c("Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","Black","orange","orange","orange","orange","orange","Blue","Blue","Blue","Blue","Blue","Blue"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3_symbiont,Strain,label=T)
ordiellipse(scores3_symbiont,Strain)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)

#
# dev.off() 

## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(CellType)),type="s",radius=0.005)

##
## interactive 3d plot, to confirm
# plot3d(scores3[,1], scores3[,2],scores3[,3],col=as.numeric(as.factor(Strain)),type="s",radius=0.005)

##
##
##

## STATISTIC TESTS - Adonis
##
Title <- paste(SampleGroup,"Adonis")
FileName <- paste("5-",Title,time,".txt")
#
ad_symbiont=adonis(t(logcounts_symbiont)~CellType+Strain, permutations = 999, data=conditions,method="euclidean")
ad_symbiont 
#
#capture.output(ad_symbiont,file=FileName)
#
Title <- paste(SampleGroup,"TukeyHSD")
FileName <- paste("5-",Title,time,".txt")
ad_symbiont=adonis(t(logcounts_symbiont)~Strain,data=conditions,method="euclidean")
ad_symbiont
pco1=scores3_symbiont[,1]
TukeyHSD(aov(pco1~Strain))
#
#capture.output(TukeyHSD(aov(pco1~Strain)),file=FileName)

```
<br><br>


## Export graphics
```{r, eval=FALSE}
# MDS plot and heatmap export

SampleGroup <-"31C"
Title <- paste(SampleGroup,"MDS-logcounts-euclidean_2")
time<-format(Sys.time(),"%d-%m-%Y")

FileName <- paste(Title,time,".jpg")
jpeg(file=FileName, width = 2020, height = 1020, res = 200)
par(mfrow=c(1,2))
#
plot(scores3_host[,1], scores3_host[,2], col="Black", xlim = c(-0.05, 0.07), ylim = c(-0.042, 0.042), bg=c("grey","grey","grey","grey","grey","grey","Black","Black","Black","Black","Black","Black","coral","coral","coral","coral","coral","Blue","Blue","Blue","Blue","Blue","Blue"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3_host,Strain,label=F)
ordiellipse(scores3_host,Strain)
#
Title <- paste(SampleGroup,"MDS-Plot")
title(Title)
plot(scores3_symbiont[,1], scores3_symbiont[,2], col="Black", xlim = c(-0.05, 0.07), ylim = c(-0.042, 0.042), bg=c("grey","grey","grey","grey","grey","grey","Black","Black","Black","Black","Black","Black","coral","coral","coral","coral","coral","Blue","Blue","Blue","Blue","Blue","Blue"),pch=c(21,21,21,21))
# plot(scores3[,1], scores3[,2],col=as.numeric(as.factor(Strain)))
ordispider(scores3_symbiont,Strain,label=F)
ordiellipse(scores3_symbiont,Strain)
#
Title <- paste(SampleGroup,"MDS-Plot")
#
title(Title)
#
dev.off() 

par(mfrow=c(1,2))

SampleGroup <-"DGE"
time<-format(Sys.time(),"%d-%m-%Y")
Title <- paste(SampleGroup,"Heat-Map-Symbiont_Photo")
FileName <- paste(Title,time,".jpg")
jpeg(file=FileName, width = 4000, height = 5500, res = 550)
par(mfrow=c(1,2))
expression_heatmap_symbiont
dev.off() 

SampleGroup <-"DGE"
time<-format(Sys.time(),"%d-%m-%Y")
Title <- paste(SampleGroup,"Heat-Map-Host_Heat")
FileName <- paste(Title,time,".jpg")
jpeg(file=FileName, width =3920, height = 4020, res = 600)
par(mfrow=c(1,2))
expression_heatmap_host_heat
dev.off() 

```
<br><br>



# __4 GOSeq gene ontology analysis__
```{r, message=FALSE, warning=FALSE, eval=FALSE}
#### Install packages
if (!requireNamespace("BiocManager", quietly = TRUE))
# install.packages("BiocManager")
# BiocManager::install("goseq", version = "3.8")
# BiocManager::install("GO.db", version = "3.8")
# BiocManager::install("qvalue", version = "3.8")
# BiocManager::install("geneLenDataBase", version = "3.8")
# BiocManager::install("geneLenDataBase", version = "3.8")
# BiocManager::install("Rsamtools", version = "3.8")

library(Rsamtools)
library(geneLenDataBase)
library(goseq)
library(GO.db)
library(qvalue)
library(treemap) 

```

***
<br><br>

## Source script
```{r, eval=FALSE}
# https://github.com/trinityrnaseq/trinity_ext_sample_data/blob/master/test_GOSeq_trinotate_pipe/Spombe/__runGOseq.R

```


## GOSeq: Coral Host 
```{r, eval=FALSE}
# capture list of genes for functional enrichment testing
####
factor_labeling = read.table("HOST/1Custom.list.HOST.txt", row.names=2, header=F)
colnames(factor_labeling) = c('type')
factor_list = unique(factor_labeling[,1])
DE_genes = rownames(factor_labeling)
#
# get gene lengths
gene_lengths = read.table("../OwnFiles/HOST/3HOST_GENE.length.txt", header=T, row.names=1)
gene_lengths = as.matrix(gene_lengths[,1,drop=F])
#
# get background gene list
background = read.table("../OwnFiles/HOST/2HOST_background_SS8_WT10", header=T, row.names=1)
background.gene_ids = rownames(background)
background.gene_ids = unique(c(background.gene_ids, DE_genes))
#
# parse GO assignments
GO_info = read.table("../OwnFiles/HOST/4HOST_blast2go_annot_20190211_1343.txt", header=F, row.names=1,stringsAsFactors=F)
#
#
GO_info_listed = apply(GO_info, 1, function(x) unlist(strsplit(x,',')))
names(GO_info_listed) = rownames(GO_info)
get_GO_term_descr =  function(x) {
    d = 'none';
    go_info = GOTERM[[x]];
    if (length(go_info) >0) { d = paste(Ontology(go_info), Term(go_info), sep=' ');}
    return(d);
}
#
# GO-Seq protocol: build pwf based on ALL DE features
sample_set_gene_ids = background.gene_ids
sample_set_gene_lengths = gene_lengths[sample_set_gene_ids,]
GO_info_listed = GO_info_listed[ names(GO_info_listed) %in% sample_set_gene_ids ]
cat_genes_vec = as.integer(sample_set_gene_ids %in% rownames(factor_labeling))
pwf=nullp(cat_genes_vec, bias.data=sample_set_gene_lengths)
rownames(pwf) = sample_set_gene_ids
#
# perform functional enrichment testing for each category.
for (feature_cat in factor_list) {
   message('Processing category: ', feature_cat)
    cat_genes_vec = as.integer(sample_set_gene_ids %in% rownames(factor_labeling)[factor_labeling$type == feature_cat])
    pwf$DEgenes = cat_genes_vec
    res = goseq(pwf,gene2cat=GO_info_listed, use_genes_without_cat=TRUE)
    ## over-represented categories:
     pvals = res$over_represented_pvalue
     pvals[pvals > 1 - 1e-10] = 1 - 1e-10
     q = qvalue(pvals)
     res$over_represented_FDR = q$qvalues
    go_enrich_filename = paste(feature_cat,'HOST.GOseq.enriched', sep='')
    result_table = res[res$over_represented_pvalue<=0.05,]
    descr = unlist(lapply(result_table$category, get_GO_term_descr))
    result_table$go_term = descr;
    write.table(result_table[order(result_table$over_represented_pvalue),], file=go_enrich_filename, sep='	', quote=F, row.names=F)
    ## under-represented categories:
     pvals = res$under_represented_pvalue
     pvals[pvals>1-1e-10] = 1 - 1e-10
     q = qvalue(pvals)
     res$under_represented_FDR = q$qvalues
    go_depleted_filename = paste(feature_cat,'HOST.GOseq.depleted', sep='')
    result_table = res[res$under_represented_pvalue<=0.05,]
    descr = unlist(lapply(result_table$category, get_GO_term_descr))
    result_table$go_term = descr;
    #
    write.table(result_table[order(result_table$under_represented_pvalue),], file=go_depleted_filename, sep='	', quote=F, row.names=F)
    #
    }

enriched.GO=res$category[p.adjust(res$over_represented_pvalue,method="BH")<.05]
head(enriched.GO)


```
<br>



## GOSeq: Algal Symbiont
```{r, eval=FALSE}
# capture list of genes for functional enrichment testing
####
factor_labeling = read.table("../OwnFiles/SYMBIONT/1Custom.list.SYMBIONT.txt", row.names=2, header=F)
colnames(factor_labeling) = c('type')
factor_list = unique(factor_labeling[,1])
DE_genes = rownames(factor_labeling)
#
# get gene lengths
gene_lengths = read.table("../OwnFiles/SYMBIONT/3Symbiont-GeneLenght.txt", header=T, row.names=1)
gene_lengths = as.matrix(gene_lengths[,1,drop=F])
#
# get background gene list
background = read.table("../OwnFiles/SYMBIONT/2Symbiont_background_SS8_VS_WT10", header=T, row.names=1)
background.gene_ids = rownames(background)
background.gene_ids = unique(c(background.gene_ids, DE_genes))
#
# parse GO assignments
GO_info = read.table("../OwnFiles/SYMBIONT/4SYMBIONT_blast2go_annot_20190211_1353.annot", header=F, row.names=1,stringsAsFactors=F)
#
GO_info_listed = apply(GO_info, 1, function(x) unlist(strsplit(x,',')))
names(GO_info_listed) = rownames(GO_info)
get_GO_term_descr =  function(x) {
    d = 'none';
    go_info = GOTERM[[x]];
    if (length(go_info) >0) { d = paste(Ontology(go_info), Term(go_info), sep=' ');}
    return(d);
}
#
# GO-Seq protocol: build pwf based on ALL DE features
sample_set_gene_ids = background.gene_ids
sample_set_gene_lengths = gene_lengths[sample_set_gene_ids,]
GO_info_listed = GO_info_listed[ names(GO_info_listed) %in% sample_set_gene_ids ]
cat_genes_vec = as.integer(sample_set_gene_ids %in% rownames(factor_labeling))
pwf=nullp(cat_genes_vec, bias.data=sample_set_gene_lengths)
rownames(pwf) = sample_set_gene_ids
#
# perform functional enrichment testing for each category.
for (feature_cat in factor_list) {
   message('Processing category: ', feature_cat)
    cat_genes_vec = as.integer(sample_set_gene_ids %in% rownames(factor_labeling)[factor_labeling$type == feature_cat])
    pwf$DEgenes = cat_genes_vec
    res = goseq(pwf,gene2cat=GO_info_listed, use_genes_without_cat=TRUE)
    ## over-represented categories:
     pvals = res$over_represented_pvalue
     pvals[pvals > 1 - 1e-10] = 1 - 1e-10
     q = qvalue(pvals)
     res$over_represented_FDR = q$qvalues
    go_enrich_filename = paste(feature_cat,'.GOseq.enriched', sep='')
    result_table = res[res$over_represented_pvalue<=0.05,]
    descr = unlist(lapply(result_table$category, get_GO_term_descr))
    result_table$go_term = descr;
    write.table(result_table[order(result_table$over_represented_pvalue),], file=go_enrich_filename, sep='	', quote=F, row.names=F)
    ## under-represented categories:
     pvals = res$under_represented_pvalue
     pvals[pvals > 1-1e-10] = 1 - 1e-10
     q = qvalue(pvals)
     res$under_represented_FDR = q$qvalues
    go_depleted_filename = paste(feature_cat,'.GOseq.depleted', sep='')
    result_table = res[res$under_represented_pvalue<=0.05,]
    descr = unlist(lapply(result_table$category, get_GO_term_descr))
    result_table$go_term = descr;
    #
    write.table(result_table[order(result_table$under_represented_pvalue),], file=go_depleted_filename, sep='	', quote=F, row.names=F)
    #
}

enriched.GO=res$category[p.adjust(res$over_represented_pvalue,method="BH")<0.05]
head(enriched.GO)

```
<br><br>



## Heat map - GO categories
```{r, eval=FALSE}

library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(BiasedUrn)
library(openxlsx)
library(vegan)
library(rgl)
library(ape)
library(openxlsx)
library(ggplot2)


highly_variable_lcpm_NEW <- read.delim("data/GO_Cats_heatmap2.txt", row.names=1)
highly_variable_lcpm_NEWNEW <- as.matrix(highly_variable_lcpm_NEW, mode='any')

#
time<-format(Sys.time(),"%d-%m-%Y")
Title <- paste("Selected Cladocopium strains vs wild type strain")
FileName <- paste(Title,time,".jpg")
#jpeg(file=FileName, width = 2220, height = 1280, res = 200)
jpeg(file=FileName, width = 2500, height = 3000, res = 220)
par(mfrow = c(1,1))
#
#
#mypalette <- brewer.pal(5, "RdGyBu")
#morecols <- colorRampPalette(mypalette)
color.palette  <- c(colorRampPalette(c("blue", "white", "red"))(n=10))
#color.palette <- (wes_palette(21, name = "Zissou1", type = "continuous")))
col.cell <- c("purple", "orange")[symbiont_samples$CellType]
colsep = "White"
rowsep = "White"
br <- seq(-7, 3, by = 0.25) 
#PVal <- read.delim("data/GO_Cats_heatmap-values.txt", row.names=1)
PVal <- read.delim("data/GO_Cats_heatmap-values.txt", row.names=1)
column_labels <- paste(colnames(highly_variable_lcpm_NEW)) #$CellType, symbiont_samples$Strain, sep = ".")
#col_breaks = c(seq(0.001,0.2,length=100), seq(0.3,0.4,length=100), seq(0.5,0.6,length=100))
#column_labels <- paste(sampleinfo$SampleName)
#
heatmap.2(
  highly_variable_lcpm_NEWNEW,
  Rowv = FALSE,
  Colv = FALSE,
  #col = rev(morecols(9)),
  #col=bluered(256),
  col = color.palette,
  #breaks = col_breaks, 
  #Colv = "NA", 
  na.color = "White",
  na.rm=TRUE,
  trace = "none",
  #cellnote=as.matrix(PVal), notecex=1.6,
  notecol="black",
  main = "GO Categories",
  density.info=c("none"), #,"density","none"),
  keysize=0.8,
  key.title = "",
  key.xlab = "Log p-value",
  key.ylab = "Counts",
  #lhei=c(1,3.5), lwid=c(1,5),
  #ColSideColors = col.cell,
  scale = "none",
  labCol = column_labels,
  ## fix stuff that falls off the image with the following line
  margins = c(8,40),
  #breaks = br
  #colsep,
  symm=F,symkey=F,symbreaks=T,
  #breaks=seq(-4, 4),
  cexRow=1.6,cexCol=1.2, 
  #rowsep,
  #rowsep=c(1, 2, 3, 4),
  colsep=c(1, 2, 3, 4, 5, 6),
  sepcolor="White"
  #sepwidth=c(0.05,0.05)
  #sepwidth=c(0.05,0.05)
  ## remove junk on side
  #labRow = ""
)
title(Title, line = +3.3)
dev.off() 






highly_variable_lcpm_HOST <- read.delim("data/GO_Cats_heatmap_host.txt", row.names=1)
highly_variable_lcpm_HOST2 <- as.matrix(highly_variable_lcpm_HOST, mode='any')

#
time<-format(Sys.time(),"%d-%m-%Y")
Title <- paste("Coral larvae with selected Cladocopium strains vs larvae with wild type")
FileName <- paste(Title,time,".jpg")
jpeg(file=FileName, width = 2500, height = 3000, res = 220)
par(mfrow = c(1,1))
#
#
#mypalette <- brewer.pal(5, "RdGyBu")
#morecols <- colorRampPalette(mypalette)
color.palette  <- c(colorRampPalette(c("blue", "lightgrey", "red"))(n=30))
col.cell <- c("purple", "orange")[symbiont_samples$CellType]
colsep = "White"
rowsep = "White"
br <- seq(-7, 3, by = 0.25) 
#PVal <- read.delim("data/GO_Cats_heatmap-values.txt", row.names=1)
PVal <- read.delim("data/GO_Cats_heatmap-values-host.txt", row.names=1)
column_labels <- paste(colnames(highly_variable_lcpm_HOST)) #$CellType, symbiont_samples$Strain, sep = ".")
#col_breaks = c(seq(0.001,0.2,length=100), seq(0.3,0.4,length=100), seq(0.5,0.6,length=100))
#column_labels <- paste(sampleinfo$SampleName)
#
heatmap.2(
  highly_variable_lcpm_HOST2,
  #Rowv = FALSE,
  #Colv = FALSE,
  #col = rev(morecols(9)),
  #col=bluered(256),
  col = color.palette,
  #breaks = col_breaks, 
  #Colv = "NA", 
  na.color = "White",
  na.rm=TRUE,
  trace = "none",
  cellnote=as.matrix(PVal), notecex=1.6,
  notecol="black",
  main = "GO Categories",
  density.info=c("none"), #,"density","none"),
  keysize=0.6,
  key.title = "",
  key.xlab = "Log p-value",
  key.ylab = "Counts",
  lhei=c(1,10), lwid=c(1,4.5),
  #ColSideColors = col.cell,
  scale = "none",
  labCol = column_labels,
  ## fix stuff that falls off the image with the following line
  margins = c(8,40),
  #breaks = br
  #colsep,
  symm=F,symkey=F,symbreaks=T,
  #breaks=seq(-4, 4),
  cexRow=1.7,cexCol=1.9, 
  #rowsep,
  rowsep=c(1:28),
  colsep=c(1, 2, 3),
  sepcolor="White"
  #sepwidth=c(0.05,0.05)
  #sepwidth=c(0.05,0.05)
  ## remove junk on side
  #labRow = ""
)
title(Title, line = +3.3)
dev.off() 




#
time<-format(Sys.time(),"%d-%m-%Y")
Title <- paste("Selected Cladocopium strains vs wild type strains")
FileName <- paste(Title,time,".jpg")
jpeg(file=FileName, width = 2500, height = 3000, res = 220)
par(mfrow = c(1,1))
#
#
#mypalette <- brewer.pal(5, "RdGyBu")
#morecols <- colorRampPalette(mypalette)
color.palette  <- c(colorRampPalette(c("blue", "lightgrey", "red"))(n=30))
col.cell <- c("purple", "orange")[symbiont_samples$CellType]
colsep = "White"
rowsep = "White"
br <- seq(-7, 3, by = 0.25) 
#PVal <- read.delim("data/GO_Cats_heatmap-values.txt", row.names=1)
PVal <- read.delim("data/GO_Cats_heatmap-values.txt", row.names=1)
column_labels <- paste(colnames(highly_variable_lcpm_NEW)) #$CellType, symbiont_samples$Strain, sep = ".")
#col_breaks = c(seq(0.001,0.2,length=100), seq(0.3,0.4,length=100), seq(0.5,0.6,length=100))
#column_labels <- paste(sampleinfo$SampleName)
#
heatmap.2(
  highly_variable_lcpm_NEWNEW,
  #Rowv = FALSE,
  #Colv = FALSE,
  #col = rev(morecols(9)),
  #col=bluered(256),
  col = color.palette,
  #breaks = col_breaks, 
  #Colv = "NA", 
  na.color = "White",
  na.rm=TRUE,
  trace = "none",
  cellnote=as.matrix(PVal), notecex=1.6,
  notecol="black",
  main = "GO Categories",
  density.info=c("none"), #,"density","none"),
  keysize=0.6,
  key.title = "",
  key.xlab = "Log p-value",
  key.ylab = "Counts",
  lhei=c(1,10), lwid=c(1,4.5),
  #ColSideColors = col.cell,
  scale = "none",
  labCol = column_labels,
  ## fix stuff that falls off the image with the following line
  margins = c(66.4,40),
  #breaks = br
  #colsep,
  symm=F,symkey=F,symbreaks=T,
  #breaks=seq(-4, 4),
  cexRow=1.7,cexCol=1.9, 
  #rowsep,
  rowsep=c(1:28),
  colsep=c(1, 2, 3),
  sepcolor="White"
  #sepwidth=c(0.05,0.05)
  #sepwidth=c(0.05,0.05)
  ## remove junk on side
  #labRow = ""
)
title(Title, line = +3.3)
dev.off() 
```
<br><br>


## Ranked based GO analysis with adaptive clustering (MWU test) 
```{r}
# Source:
# https://github.com/z0on/GO_MWU
```


<br><br>








